# Intro notebook
The purpose of this notebook is to connect to a local database that we created by loading CSV files into that database.
Copy / paste this notebook into other places to add / delete cells and add new analysis.

In [2]:
import os
print(os.getcwd())

/Users/beoconno/Documents/ff/ff_data/notebooks/nfl


In [3]:
from pathlib import Path

In [4]:
%load_ext sql

In [5]:
# Define your database file path (SQLite example)
rel_db_path = Path.cwd().parent.parent / 'db_output' / 'nfl.db'
db_file_path = rel_db_path.resolve()
db_string = f"sqlite:///{db_file_path}"
%sql $db_string

In [104]:
%%sql
select
   COUNT(DISTINCT team_id) AS team_cnt
from image_roster_1combo_eval
limit 10 
;

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[(92154,)]

There are 92,154 teams total.
But a lot of them don't end up playing people on their roster because they selected better players.
Once we look at those combos, there are only 10,652 distinct rosters.

In [107]:
%%sql

SELECT
   COUNT(DISTINCT roster) AS roster_cnt
FROM (
   SELECT
      team_id,
      GROUP_CONCAT(iif(team_score > 0, name,"")) AS roster
   from (SELECT * FROM image_roster_1combo_eval ORDER BY CAST(round AS INT) ASC)
   GROUP BY 1
)
limit 10 
;

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[(10652,)]

In [117]:
%%sql
SELECT
   roster,
   AVG(team_total) AS team_pts,
   MIN(team_id) AS example_team
FROM (
   SELECT
      team_id,
      GROUP_CONCAT(iif(team_score > 0, name,"")) AS roster,
      SUM(team_score) AS team_total
   from (SELECT * FROM image_roster_1combo_eval ORDER BY CAST(round AS INT) ASC)
   GROUP BY 1
)
-- WHERE roster NOT LIKE 'christian mccaffrey%'
GROUP BY 1
ORDER BY 2 DESC
limit 50 
;

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,dj moore,,alvin kamara,evan engram,', 141.04, 9579),
 ('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,,mike williams,alvin kamara,evan engram,', 140.92000000000004, 9523),
 ('christian mccaffrey,ceedee lamb,josh allen,deebo samuel,keenan allen,,,alvin kamara,evan engram,', 140.44000000000003, 10275),
 ('christian mccaffrey,ceedee lamb,josh allen,jahmyr gibbs,keenan allen,,,alvin kamara,evan engram,', 140.32999999999996, 9891),
 ('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,dj moore,,alvin kamara,,david njoku', 140.07000000000005, 37225),
 ('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,,rachaad white,alvin kamara,evan engram,', 139.95000000000002, 9507),
 ('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,,mike williams,alvin kamara,,david njoku', 139.94999999999993, 37169),
 ('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,james conner,,alvin kamara,evan engram,', 139.69, 9539),
 ('christian mccaffrey,ceedee lamb,josh allen,deebo samuel,keenan allen,,,alvin kamara,,david njoku', 139.46999999999997, 37921),
 ('christian mccaffrey,ceedee lamb,josh allen,jahmyr gibbs,keenan allen,,,alvin kamara,,david njoku', 139.36000000000007, 37537),
 ('christian mccaffrey,ceedee lamb,josh allen,jahmyr gibbs,keenan allen,,mike williams,,evan engram,', 139.14000000000001, 9983),
 ('christian mccaffrey,ceedee lamb,josh allen,jahmyr gibbs,keenan allen,,,mike evans,evan engram,', 139.03000000000003, 9911),
 ('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,,rachaad white,alvin kamara,,david njoku', 138.98000000000002, 37153),
 ('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,james conner,,alvin kamara,,david njoku', 138.72, 37185),
 ('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,,rachaad white,mike evans,evan engram,', 138.65, 9583),
 ('christian mccaffrey,ceedee lamb,,lamar jackson,keenan allen,dj moore,,alvin kamara,evan engram,', 138.64999999999998, 10731),
 ('christian mccaffrey,ceedee lamb,mark andrews,lamar jackson,keenan allen,dj moore,,alvin kamara,,', 138.64, 11881),
 ('christian mccaffrey,ceedee lamb,,lamar jackson,keenan allen,,mike williams,alvin kamara,evan engram,', 138.53000000000003, 10675),
 ('christian mccaffrey,ceedee lamb,mark andrews,lamar jackson,keenan allen,,mike williams,alvin kamara,,', 138.5200000000001, 11825),
 ('christian mccaffrey,ceedee lamb,travis etienne,lamar jackson,keenan allen,,,alvin kamara,evan engram,', 138.40999999999997, 10659),
 ('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,james conner,,mike evans,evan engram,', 138.39000000000001, 9551),
 ('christian mccaffrey,ceedee lamb,josh allen,deebo samuel,keenan allen,,rachaad white,,evan engram,', 138.28, 10279),
 ('christian mccaffrey,ceedee lamb,josh allen,jahmyr gibbs,keenan allen,,mike williams,,,david njoku', 138.17000000000002, 37629),
 ('christian mccaffrey,ceedee lamb,mark andrews,,keenan allen,dj moore,,alvin kamara,dak prescott,', 138.09999999999994, 12265),
 ('christian mccaffrey,ceedee lamb,josh allen,jahmyr gibbs,keenan allen,,,mike evans,,david njoku', 138.05999999999997, 37557),
 ('christian mccaffrey,ceedee lamb,josh allen,deebo samuel,keenan allen,james conner,,,evan engram,', 138.02, 10311),
 ('christian mccaffrey,ceedee lamb,mark andrews,,keenan allen,,mike williams,alvin kamara,dak prescott,', 137.98000000000002, 12209),
 ('christian mccaffrey,ceedee lamb,josh allen,,keenan allen,,rachaad white,mike evans,,david njoku', 137.68000000000004, 37229),
 ('christian mccaffrey,ceedee lamb,,lamar jackson,keenan allen,dj moore,,alvin kamara,,david njoku', 137.68, 38377),
 ('christian mccaffrey,ceedee lamb,,lamar jackson,keenan allen,,rachaad white,alvin kamara,evan engram,', 137.55999999999997, 11811),
 ('christian mccaffrey,ceedee lamb,,lamar jackson,keenan allen,,mike williams,alvin kamara,,david njoku', 137.55999999999997, 38321),
 ('christian mccaffrey,ceedee lamb,mark andrews,lamar jackson,

In [119]:
%%sql
SELECT
    *,
    SUM(team_score) OVER(PARTITION BY team_id) AS team_total
FROM image_roster_1combo_eval
WHERE 1=1
    AND (
        team_id = 119979
        OR team_id = 9579
    )
LIMIT 20

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('josh allen', 'QB', 17, 23.09, '3', '2023', 9579, 1, 9, 3, 3, 3, 1, 23.09, 141.04),
 ('lamar jackson', 'QB', 16, 20.7, '4', '2023', 9579, 2, 10, 3, 3, 3, 1, 0, 141.04),
 ('tua tagovailoa', 'QB', 17, 15.91, '10', '2023', 9579, 3, 8, 3, 3, 3, 1, 0, 141.04),
 ('christian mccaffrey', 'RB', 16, 24.46, '1', '2023', 9579, 1, 1, 3, 3, 3, 1, 24.46, 141.04),
 ('alvin kamara', 'RB', 13, 17.92, '8', '2023', 9579, 2, 4, 3, 3, 3, 1, 17.92, 141.04),
 ('rachaad white', 'RB', 17, 15.76, '7', '2023', 9579, 3, 6, 3, 3, 3, 1, 0, 141.04),
 ('evan engram', 'TE', 17, 13.55, '9', '2023', 9579, 1, 7, 3, 3, 3, 1, 13.55, 141.04),
 ('ceedee lamb', 'WR', 17, 23.72, '2', '2023', 9579, 1, 2, 3, 3, 3, 1, 23.72, 141.04),
 ('keenan allen', 'WR', 13, 21.45, '5', '2023', 9579, 2, 3, 3, 3, 3, 1, 21.45, 141.04),
 ('dj moore', 'WR', 17, 16.85, '6', '2023', 9579, 3, 5, 3, 3, 3, 1, 16.85, 141.04),
 ('josh allen', 'QB', 17, 23.09, '3', '2023', 119979, 1, 9, 3, 3, 3, 1, 23.09, 135.75),
 ('lamar jackson', 'QB', 16, 20.7, '4', '2023', 119979, 2, 10, 3, 3, 3, 1, 0, 135.75),
 ('tua tagovailoa', 'QB', 17, 15.91, '10', '2023', 119979, 3, 8, 3, 3, 3, 1, 0, 135.75),
 ('alvin kamara', 'RB', 13, 17.92, '8', '2023', 119979, 1, 3, 3, 3, 3, 1, 17.92, 135.75),
 ('breece hall', 'RB', 17, 17.09, '5', '2023', 119979, 2, 4, 3, 3, 3, 1, 17.09, 135.75),
 ('rachaad white', 'RB', 17, 15.76, '7', '2023', 119979, 3, 6, 3, 3, 3, 1, 0, 135.75),
 ('evan engram', 'TE', 17, 13.55, '9', '2023', 119979, 1, 7, 3, 3, 3, 1, 13.55, 135.75),
 ('ceedee lamb', 'WR', 17, 23.72, '2', '2023', 119979, 1, 1, 3, 3, 3, 1, 23.72, 135.75),
 ('tyreek hill', 'WR', 16, 23.53, '1', '2023', 119979, 2, 2, 3, 3, 3, 1, 23.53, 135.75),
 ('dj moore', 'WR', 17, 16.85, '6', '2023', 119979, 3, 5, 3, 3, 3, 1, 16.85, 135.75)]

In [131]:
%%sql
SELECT
    round,
    name,
    position,
    AVG(team_total) AS avg_total,
    MAX(team_total) AS max_total,
    MIN(team_total) AS min_total,
    COUNT(1) AS rosters_made
FROM (
    SELECT
        round,
        name,
        position,
        year,
        SUM(team_score) OVER(PARTITION BY team_id) AS team_total
    FROM image_roster_1combo_eval
    WHERE 1=1
        AND team_score > 0
    )
WHERE 1=1
    AND team_total > 120
GROUP BY 1, 2, 3
ORDER BY COUNT(1) DESC

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('8', 'alvin kamara', 'RB', 127.58783448275882, 141.04, 120.01, 18850),
 ('1', 'christian mccaffrey', 'RB', 128.09603639392557, 141.04, 120.01, 15607),
 ('2', 'ceedee lamb', 'WR', 129.6561720372492, 141.04, 120.03999999999999, 14927),
 ('1', 'tyreek hill', 'WR', 126.19330688336593, 135.75, 120.01, 10460),
 ('8', 'mike evans', 'WR', 125.89988397389503, 139.03, 120.01, 9653),
 ('5', 'breece hall', 'RB', 126.31456728145928, 135.75, 120.01, 9163),
 ('7', 'mike williams', 'WR', 127.05767681191197, 140.92, 120.02000000000002, 9134),
 ('4', 'deebo samuel', 'WR', 126.06714620535708, 140.44, 120.01, 8960),
 ('5', 'keenan allen', 'WR', 128.47838267480682, 141.04, 120.02000000000001, 8681),
 ('4', 'jahmyr gibbs', 'RB', 126.43769966523317, 140.32999999999998, 120.02000000000002, 8364),
 ('3', 'josh allen', 'QB', 128.87602141482154, 141.04, 120.03, 8312),
 ('3', 'travis etienne', 'RB', 125.86434029546744, 138.41, 120.02000000000001, 7852),
 ('7', 'rachaad white', 'RB', 126.40421348314742, 139.95, 120.01, 7832),
 ('6', 'dj moore', 'WR', 127.14765900942872, 141.04, 120.04, 7531),
 ('4', 'lamar jackson', 'QB', 127.55119629430972, 138.65, 120.02000000000001, 7448),
 ('9', 'evan engram', 'TE', 127.7861868256672, 141.04, 120.01, 7044),
 ('9', 'dak prescott', 'QB', 127.17091698113286, 138.10000000000002, 120.03, 5300),
 ('6', 'james conner', 'RB', 125.94359576515915, 139.69, 120.01, 5195),
 ('5', 'tj hockenson', 'TE', 127.27002122323009, 137.36999999999998, 120.03999999999999, 5183),
 ('3', 'mark andrews', 'TE', 127.2568560050299, 138.64000000000001, 120.02000000000002, 4771),
 ('10', 'david njoku', 'TE', 127.66072144288727, 140.07, 120.01, 4491),
 ('1', 'travis kelce', 'TE', 124.78233699462622, 133.68, 120.07000000000001, 4279),
 ('2', 'derrick henry', 'RB', 123.74471930906888, 131.83, 120.02000000000001, 3242),
 ('7', 'dallas goedert', 'TE', 126.10601510859284, 137.23, 120.08, 3177),
 ('2', 'patrick mahomes', 'QB', 123.28087633885077, 128.47, 120.01, 3081),
 ('5', 'justin herbert', 'QB', 124.68149507948588, 131.25, 120.02000000000001, 2642),
 ('3', 'chris olave', 'WR', 124.04394657258122, 136.26, 120.02000000000001, 1984),
 ('6', 'trevor lawrence', 'QB', 126.16790195044719, 134.24, 120.03000000000002, 1897),
 ('6', 'darren waller', 'TE', 126.24122033898338, 136.8, 120.03, 1770),
 ('10', 'tua tagovailoa', 'QB', 125.64129051620608, 133.86, 120.01, 1666),
 ('10', 'brian robinson', 'RB', 125.48693409742148, 136.21, 120.05000000000003, 1396),
 ('9', 'jordan addison', 'WR', 123.95755458515276, 136.24, 120.03999999999999, 1145),
 ('10', 'brandin cooks', 'WR', 123.32376012965987, 131.21, 120.05999999999999, 617),
 ('9', 'khalil herbert', 'RB', 123.96032967032946, 131.41, 120.02000000000001, 546)]

In [102]:
%%sql
select
    team_id,
    GROUP_CONCAT(round || ': ' || name || '(' || team_score || ')
    ') AS roster,
    SUM(team_score) AS total_pts
from (SELECT * FROM image_roster_1combo_eval ORDER BY CAST(round AS INT) ASC)
group by team_id
order by 3 desc
limit 10 
;

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[(51835, '1: christian mccaffrey(24.46)\n    ,2: ceedee lamb(23.72)\n    ,3: josh allen(23.09)\n    ,4: deebo samuel(0)\n    ,5: keenan allen(21.45)\n    ,6: dj moore(16.85)\n    ,7: mike williams(0)\n    ,8: alvin kamara(17.92)\n    ,9: evan engram(13.55)\n    ,10: brandin cooks(0)\n    ', 141.04),
 (51827, '1: christian mccaffrey(24.46)\n    ,2: ceedee lamb(23.72)\n    ,3: josh allen(23.09)\n    ,4: deebo samuel(0)\n    ,5: keenan allen(21.45)\n    ,6: dj moore(16.85)\n    ,7: dallas goedert(0)\n    ,8: alvin kamara(17.92)\n    ,9: evan engram(13.55)\n    ,10: brandin cooks(0)\n    ', 141.04),
 (51819, '1: christian mccaffrey(24.46)\n    ,2: ceedee lamb(23.72)\n    ,3: josh allen(23.09)\n    ,4: deebo samuel(0)\n    ,5: keenan allen(21.45)\n    ,6: dj moore(16.85)\n    ,7: rachaad white(0)\n    ,8: alvin kamara(17.92)\n    ,9: evan engram(13.55)\n    ,10: brandin cooks(0)\n    ', 141.04),
 (51451, '1: christian mccaffrey(24.46)\n    ,2: ceedee lamb(23.72)\n    ,3: josh allen(23.09)\n    ,4: jahmyr gibbs(0)\n    ,5: keenan allen(21.45)\n    ,6: dj moore(16.85)\n    ,7: mike williams(0)\n    ,8: alvin kamara(17.92)\n    ,9: evan engram(13.55)\n    ,10: brandin cooks(0)\n    ', 141.04),
 (51443, '1: christian mccaffrey(24.46)\n    ,2: ceedee lamb(23.72)\n    ,3: josh allen(23.09)\n    ,4: jahmyr gibbs(0)\n    ,5: keenan allen(21.45)\n    ,6: dj moore(16.85)\n    ,7: dallas goedert(0)\n    ,8: alvin kamara(17.92)\n    ,9: evan engram(13.55)\n    ,10: brandin cooks(0)\n    ', 141.04),
 (51435, '1: christian mccaffrey(24.46)\n    ,2: ceedee lamb(23.72)\n    ,3: josh allen(23.09)\n    ,4: jahmyr gibbs(0)\n    ,5: keenan allen(21.45)\n    ,6: dj moore(16.85)\n    ,7: rachaad white(0)\n    ,8: alvin kamara(17.92)\n    ,9: evan engram(13.55)\n    ,10: brandin cooks(0)\n    ', 141.04),
 (51067, '1: christian mccaffrey(24.46)\n    ,2: ceedee lamb(23.72)\n    ,3: josh allen(23.09)\n    ,4: lamar jackson(0)\n    ,5: keenan allen(21.45)\n    ,6: dj moore(16.85)\n    ,7: mike williams(0)\n    ,8: alvin kamara(17.92)\n    ,9: evan engram(13.55)\n    ,10: brandin cooks(0)\n    ', 141.04),
 (51059, '1: christian mccaffrey(24.46)\n    ,2: ceedee lamb(23.72)\n    ,3: josh allen(23.09)\n    ,4: lamar jackson(0)\n    ,5: keenan allen(21.45)\n    ,6: dj moore(16.85)\n    ,7: dallas goedert(0)\n    ,8: alvin kamara(17.92)\n    ,9: evan engram(13.55)\n    ,10: brandin cooks(0)\n    ', 141.04),
 (51051, '1: christian mccaffrey(24.46)\n    ,2: ceedee lamb(23.72)\n    ,3: josh allen(23.09)\n    ,4: lamar jackson(0)\n    ,5: keenan allen(21.45)\n    ,6: dj moore(16.85)\n    ,7: rachaad white(0)\n    ,8: alvin kamara(17.92)\n    ,9: evan engram(13.55)\n    ,10: brandin cooks(0)\n    ', 141.04),
 (38011, '1: christian mccaffrey(24.46)\n    ,2: ceedee lamb(23.72)\n    ,3: josh allen(23.09)\n    ,4: deebo samuel(0)\n    ,5: keenan allen(21.45)\n    ,6: dj moore(16.85)\n    ,7: mike williams(0)\n    ,8: alvin kamara(17.92)\n    ,9: evan engram(13.55)\n    ,10: david njoku(0)\n    ', 141.04)]

In [90]:
%%sql
SELECT
    team_id,
    year,
    GROUP_CONCAT(round || ': ' || name) AS roster,
    SUM(IIF(position = 'QB', 1, 0)) AS qb_cnt,
    SUM(IIF(position = 'RB', 1, 0)) AS rb_cnt,
    SUM(IIF(position = 'WR', 1, 0)) AS wr_cnt,
    SUM(IIF(position = 'TE', 1, 0)) AS te_cnt
FROM draft_combos_dr10
GROUP BY 1, 2
HAVING SUM(IIF(position = 'QB', 1, 0)) > 1
    AND SUM(IIF(position = 'RB', 1, 0)) >= 2
    AND SUM(IIF(position = 'WR', 1, 0)) >= 2
    AND SUM(IIF(position = 'TE', 1, 0)) >= 1
ORDER BY 4 ASC
LIMIT 20


 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[(1672, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: james conner,7: dallas goedert,8: mike evans,9: jordan addison', 2, 5, 2, 1),
 (1679, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: james conner,7: mike williams,8: mike evans,9: evan engram', 2, 5, 2, 1),
 (1688, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: darren waller,7: rachaad white,8: mike evans,9: jordan addison', 2, 5, 2, 1),
 (1696, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: darren waller,7: dallas goedert,8: mike evans,9: jordan addison', 2, 4, 2, 2),
 (1700, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: darren waller,7: mike williams,8: alvin kamara,9: jordan addison', 2, 5, 2, 1),
 (1702, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: darren waller,7: mike williams,8: mike evans,9: khalil herbert', 2, 5, 2, 1),
 (1703, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: darren waller,7: mike williams,8: mike evans,9: evan engram', 2, 4, 2, 2),
 (1704, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: darren waller,7: mike williams,8: mike evans,9: jordan addison', 2, 4, 3, 1),
 (1711, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: dj moore,7: rachaad white,8: mike evans,9: evan engram', 2, 5, 2, 1),
 (1716, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: dj moore,7: dallas goedert,8: alvin kamara,9: jordan addison', 2, 5, 2, 1),
 (1718, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: dj moore,7: dallas goedert,8: mike evans,9: khalil herbert', 2, 5, 2, 1),
 (1719, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: dj moore,7: dallas goedert,8: mike evans,9: evan engram', 2, 4, 2, 2),
 (1720, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: dj moore,7: dallas goedert,8: mike evans,9: jordan addison', 2, 4, 3, 1),
 (1723, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: dj moore,7: mike williams,8: alvin kamara,9: evan engram', 2, 5, 2, 1),
 (1727, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: breece hall,6: dj moore,7: mike williams,8: mike evans,9: evan engram', 2, 4, 3, 1),
 (1760, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: tj hockenson,6: james conner,7: rachaad white,8: mike evans,9: jordan addison', 2, 5, 2, 1),
 (1768, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: tj hockenson,6: james conner,7: dallas goedert,8: mike evans,9: jordan addison', 2, 4, 2, 2),
 (1772, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: tj hockenson,6: james conner,7: mike williams,8: alvin kamara,9: jordan addison', 2, 5, 2, 1),
 (1774, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis etienne,4: jahmyr gibbs,5: tj hockenson,6: james conner,7: mike williams,8: mike evans,9: khalil herbert', 2, 5, 2, 1),
 (1775, '2023', '1: christian mccaffrey,10: tua tagovailoa,2: patrick mahomes,3: travis 

In [36]:
%%sql
WITH prep AS (
    SELECT
        *,
        ROW_NUMBER() OVER(PARTITION BY team_id, position ORDER BY fppg_ppr DESC) AS pos_rank,
        CASE
            WHEN position IN ("QB", "TE") AND
                ROW_NUMBER() OVER(PARTITION BY team_id, position ORDER BY fppg_ppr DESC) = 1
                THEN fppg_ppr
            WHEN position IN ("RB", "WR") AND
                ROW_NUMBER() OVER(PARTITION BY team_id, position ORDER BY fppg_ppr DESC) <= 2
                THEN fppg_ppr
            ELSE 0
            END AS pre_flex
    FROM draft_combos_dr10
    ),
with_flex AS (
    SELECT
        *,
        ROW_NUMBER() OVER(
            PARTITION BY team_id
            ORDER BY IIF(pre_flex=0 AND position IN ("WR", "RB", "TE"),fppg_ppr,0) DESC
            )
    FROM prep
)
SELECT *
FROM with_flex
WHERE team_id = 44010

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('josh allen', 'QB', 17, 23.09, '3', '2023', 44010, 1, 23.09, 1),
 ('lamar jackson', 'QB', 16, 20.7, '4', '2023', 44010, 2, 0, 2),
 ('christian mccaffrey', 'RB', 16, 24.46, '1', '2023', 44010, 1, 24.46, 3),
 ('ceedee lamb', 'WR', 17, 23.72, '2', '2023', 44010, 1, 23.72, 4),
 ('keenan allen', 'WR', 13, 21.45, '5', '2023', 44010, 2, 21.45, 5)]

In [39]:
%%sql
WITH prep AS (
    SELECT
        *,
        ROW_NUMBER() OVER(PARTITION BY team_id, position ORDER BY fppg_ppr DESC) AS pos_rank,
        CASE
            WHEN position IN ("QB", "TE") AND
                ROW_NUMBER() OVER(PARTITION BY team_id, position ORDER BY fppg_ppr DESC) = 1
                THEN fppg_ppr
            WHEN position IN ("RB", "WR") AND
                ROW_NUMBER() OVER(PARTITION BY team_id, position ORDER BY fppg_ppr DESC) <= 2
                THEN fppg_ppr
            ELSE 0
            END AS pre_flex
    FROM draft_combos_dr10
    ),
with_flex AS (
    SELECT
        *,
        IIF(
            pre_flex=0 AND position IN ("WR", "RB", "TE"),
            ROW_NUMBER() OVER(
                PARTITION BY team_id
                ORDER BY IIF(pre_flex=0 AND position IN ("WR", "RB", "TE"),fppg_ppr,0) DESC),0
            ) AS flex_score
    FROM prep
)
SELECT
    team_id,
    SUM(IIF(team_score) AS total_score,
    group_concat(IIF(team_score > 0, name, ""))
FROM with_flex
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('josh allen', 'QB', 17, 23.09, '3', '2023', 44010, 1, 23.09, 0),
 ('lamar jackson', 'QB', 16, 20.7, '4', '2023', 44010, 2, 0, 0),
 ('christian mccaffrey', 'RB', 16, 24.46, '1', '2023', 44010, 1, 24.46, 0),
 ('ceedee lamb', 'WR', 17, 23.72, '2', '2023', 44010, 1, 23.72, 0),
 ('keenan allen', 'WR', 13, 21.45, '5', '2023', 44010, 2, 21.45, 0)]

In [273]:
%%sql
SELECT
    c_year AS year,
    c_pos AS position,
    CASE
        WHEN adp_pos_rank <= 3 THEN '1-3'
        WHEN adp_pos_rank <= 10 THEN '4-10'
        ELSE
            CAST((((adp_pos_rank - 1) / 10) * 10 + 1) AS VARCHAR)
            || '-' ||
            CAST((((adp_pos_rank - 1) / 10) * 10 + 10) AS VARCHAR)
        END AS adp_bucket,
    MIN(adp_pos_rank) AS bucket_start,
    COUNT(1) AS player_cnt,
    ROUND(AVG(exp_fppg_ppr),2) AS avg_expected_ppr,
    ROUND(AVG(fppg_ppr),2) AS avg_actual_ppr,
    ROUND(AVG(pick_accuracy),2) AS avg_accuracy,
    MAX(pick_accuracy) AS biggest_surprise,
    MIN(pick_accuracy) AS worst_miss
FROM nfl_adp_hype
WHERE 1=1
    AND fppg_ppr IS NOT NULL
    AND adp_pos_rank IS NOT NULL
    AND exp_fppg_ppr IS NOT NULL
GROUP BY 1, 2, 3
ORDER BY adp_pos_rank
LIMIT 20

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
   sqlite:////Users/beoconno/Documents/ff/ff_data/raw_data/nfl/src_code/nfl.db
Done.


[('2016', 'QB', '1-3', 1, 3, 23.32, 19.19, -4.12, 1.13, -8.14),
 ('2016', 'RB', '1-3', 1, 3, 23.33, 19.86, -3.48, 2.66, -13.7),
 ('2016', 'WR', '1-3', 1, 3, 21.48, 19.2, -2.28, -1.49, -2.71),
 ('2017', 'QB', '1-3', 1, 3, 23.32, 17.8, -5.51, -4.13, -6.58),
 ('2017', 'RB', '1-3', 1, 3, 23.33, 17.42, -5.92, -0.06, -13.09),
 ('2018', 'QB', '1-3', 1, 3, 23.32, 19.28, -4.03, -1.89, -5.55),
 ('2018', 'TE', '1-3', 1, 3, 15.68, 15.34, -0.34, 3.12, -6.96),
 ('2018', 'WR', '1-3', 1, 3, 21.48, 20.54, -0.94, -0.37, -1.62),
 ('2019', 'RB', '1-3', 1, 3, 23.33, 21.99, -1.34, 8.37, -7.31),
 ('2019', 'TE', '1-3', 1, 3, 15.68, 15.38, -0.3, 0.3, -1.16),
 ('2019', 'WR', '1-3', 1, 3, 21.48, 19.68, -1.8, 3.38, -5.3),
 ('2020', 'WR', '1-3', 1, 3, 21.48, 19.84, -1.64, 4.39, -11.21),
 ('2021', 'QB', '1-3', 1, 3, 23.32, 22.14, -1.18, 1.06, -3.81),
 ('2021', 'RB', '1-3', 1, 3, 23.33, 17.38, -5.96, -3.03, -7.88),
 ('2022', 'RB', '1-3', 1, 3, 23.33, 18.73, -4.6, 0.84, -12.78),
 ('2022', 'WR', '1-3', 1, 3, 21.48, 21.42, -0.06, 0.48, -0.82),
 ('2023', 'QB', '1-3', 1, 3, 23.32, 20.53, -2.79, 0.47, -7.58),
 ('2023', 'TE', '1-3', 1, 3, 15.68, 14.26, -1.42, 0.2, -2.42),
 ('2016', 'TE', '1-3', 1, 4, 17.47, 10.05, -7.42, -1.44, -21.77),
 ('2017', 'WR', '1-3', 1, 3, 21.48, 18.8, -2.68, -1.04, -5.47)]

In [206]:
%%sql
SELECT
    c_year AS year,
    name || " (" 
        || c_pos || adp_pos_rank || " -> " ||
        c_pos || actual_pos_fin_fppg_ppr || ")" AS label,
    c_pos AS position,
    adp_pos_rank AS draft_rank,
    exp_fppg_ppr AS expected_ppg,
    fppg_ppr AS actual_ppg,
    pick_accuracy,
    CASE
        WHEN adp_pos_rank = actual_pos_fin_fppg_ppr THEN "lightgrey"
        WHEN ABS(pick_accuracy) < 2 THEN "lightgrey"
        WHEN CAST(adp_pos_rank AS INT) > CAST(actual_pos_fin_fppg_ppr AS INT) THEN "lightgreen"
        WHEN CAST(adp_pos_rank AS INT) < CAST(actual_pos_fin_fppg_ppr AS INT) THEN "lightcoral"
        ELSE "untagged"
        END AS accuracy_color
FROM nfl_adp_hype
WHERE 1=1
    AND c_pos IN ("QB", "WR", "RB", "TE")
    AND name LIKE 'antonio brown'
ORDER BY c_year ASC
LIMIT 10

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
   sqlite:////Users/beoconno/Documents/ff/ff_data/raw_data/nfl/src_code/nfl.db
Done.


[('2016', 'antonio brown (WR1 -> WR1)', 'WR', 1, 23.2, 20.49, -2.71, 'lightgrey'),
 ('2017', 'antonio brown (WR1 -> WR1)', 'WR', 1, 23.2, 22.16, -1.04, 'lightgrey'),
 ('2018', 'antonio brown (WR1 -> WR2)', 'WR', 1, 23.2, 21.58, -1.62, 'lightgrey'),
 ('2019', 'antonio brown (WR9 -> WR9)', 'WR', 9, 16.82, 16.1, -0.72, 'lightgrey'),
 ('2020', 'antonio brown (WR71 -> WR23)', 'WR', 71, 8.16, 14.64, 6.48, 'lightcoral'),
 ('2021', 'antonio brown (WR29 -> WR8)', 'WR', 29, 13.13, 17.3, 4.17, 'lightcoral')]

In [250]:
%%sql
SELECT
    year,
    position,
    label
FROM image_ep_1_accuracy_graph
WHERE 1=1
    AND LABEL != ''
LIMIT 10

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
   sqlite:////Users/beoconno/Documents/ff/ff_data/raw_data/nfl/src_code/nfl.db
Done.


[('2016', 'QB', 'cam newton<br>QB1 to 13 (-8.14) ppg'),
 ('2016', 'QB', 'matt ryan<br>QB19 to 2 (5.66) ppg'),
 ('2016', 'QB', 'tony romo<br>QB28 to 48 (-8.64) ppg'),
 ('2016', 'QB', 'mark sanchez<br>QB39 to 68 (-10.7) ppg'),
 ('2016', 'QB', 'matt moore<br>QB58 to 27 (11.11) ppg'),
 ('2016', 'QB', 'jacoby brissett<br>QB78 to 38 (9.58) ppg'),
 ('2016', 'RB', 'todd gurley<br>RB1 to 24 (-13.7) ppg'),
 ('2016', 'RB', 'adrian peterson<br>RB4 to 95 (-16.94) ppg'),
 ('2016', 'RB', 'jamaal charles<br>RB8 to 78 (-13.05) ppg'),
 ('2016', 'RB', 'jordan howard<br>RB62 to 9 (8.98) ppg')]

In [135]:
%%sql
SELECT *
FROM nfl_adp_hype
WHERE c_year = '2024'
LIMIT 10
;

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('Kyler Murray', 71.0, 'ARI', 'QB8', 'QB', 8, '2024', 72.0, 1.0, '8', 'kyler murray', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'kyler murray', 'Kyler Murray', '2024', 'ARI', 'QB', '2019', '2023', 6, None, None, 18.25, 18.3, 'ARI', None, 'ARI', 344, 344, 344, 344, 8, 19.28, 0.98, 0, '', 1, 1, 1),
 ('Kirk Cousins', 123.0, 'ATL', 'QB18', 'QB', 18, '2024', 122.7, 3.0, '13', 'kirk cousins', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'kirk cousins', 'Kirk Cousins', '2024', 'ATL', 'QB', '2016', '2023', 9, None, None, 18.75, 18.71, 'MIN', None, 'ATL', 396, 396, 396, 396, 8, 16.31, -2.4, 0, '', 1, 1, 1),
 ('Michael Penix Jr.', 264.0, 'ATL', 'QB37', 'QB', 37, '2024', 246.5, 4.0, '27', 'michael penix', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'michael penix', 'Michael Penix Jr.', '2024', 'ATL', 'QB', None, None, 1, None, None, 0, 0, None, None, 'ATL', 522, 522, 522, 522, 8, 11.17, 11.17, 0, '', 2, 2, 2),
 ('Lamar Jackson', 41.0, 'BAL', 'QB4', 'QB', 4, '2024', 42.0, 1.0, '5', 'lamar jackson', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'lamar jackson', 'Lamar Jackson', '2024', 'BAL', 'QB', '2018', '2023', 7, None, None, 20.69, 20.7, 'BAL', None, 'BAL', 314, 314, 314, 314, 8, 21.29, 0.59, 0, '', 1, 1, 1),
 ('Josh Allen', 22.0, 'BUF', 'QB1', 'QB', 1, '2024', 25.0, 2.0, '3', 'josh allen', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'josh allen', 'Josh Allen', '2024', 'BUF', 'QB', '2018', '2023', 7, None, None, 23.12, 23.09, 'BUF', None, 'BUF', 295, 295, 295, 295, 8, 25.09, 2.0, 0, '', 1, 1, 1),
 ('Bryce Young', 176.0, 'CAR', 'QB26', 'QB', 26, '2024', 178.0, 6.0, '18', 'bryce young', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'bryce young', 'Bryce Young', '2024', 'CAR', 'QB', '2023', '2023', 2, None, None, 9.75, 9.78, 'CAR', None, 'CAR', 449, 449, 449, 449, 8, 14.3, 4.52, 0, '', 1, 1, 1),
 ('Caleb Williams', 108.0, 'CHI', 'QB14', 'QB', 14, '2024', 105.7, 8.0, '11', 'caleb williams', None, None, None, None, None, None, None, None, None, None, None, None, None, None, Non

In [141]:
%%sql
SELECT
    COUNT(1)
FROM (
    SELECT
        *,
        -- It is never optimal to select anything but the best player for that position in that round
        -- This dramaticaly reduces the number of possibilities that we have to consider
        ROW_NUMBER() OVER(PARTITION BY c_year, c_pos, dr10 ORDER BY exp_fppg_ppr DESC) AS best_in_rnd
    FROM nfl_adp_hype
    )
WHERE 1=1
    AND c_pos IN ("QB", "WR", "RB", "TE")
    AND dr10 IS NOT NULL
    AND CAST(dr10 AS INT) <= 10
    AND c_year = '2024'
    AND best_in_rnd = 1
ORDER BY CAST(dr10 AS INT) ASC
LIMIT 10
;

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[(35,)]

In [161]:
%%sql
WITH rosters AS (
SELECT
    team_id,
    GROUP_CONCAT(IIF(team_score > 0, name, "")) AS team,
    SUM(team_score) AS team_total
FROM (
    SELECT *
    FROM image_roster_1combo_eval_2024
    WHERE 1=1
    ORDER BY CAST(round AS INT) ASC
    )
GROUP BY 1
)
SELECT
    team,
    AVG(team_total),
    GROUP_CONCAT(team_id) AS team_list
FROM rosters
GROUP BY 1
ORDER BY 2 DESC
LIMIT 20

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('christian mccaffrey,jahmyr gibbs,josh allen,stefon diggs,trey mcbride,george pickens,aaron jones,,,', 127.22999999999986, '2845,2846,2847,2848,2849,2850,2851,2852,2853,2854,2855,2856,27421,27422,27423,27424,27425,27426,27427,27428,27429,27430,27431,27432,51997,51998,51999,52000,52001,52002,52003,52004,52005,52006,52007,52008,76573,76574,76575,76576,76577,76578,76579,76580,76581,76582,76583,76584'),
 ('christian mccaffrey,jahmyr gibbs,josh allen,travis kelce,malik nabers,,aaron jones,deandre hopkins,,', 126.36999999999996, '2182,2183,2184,26758,26759,26760,51334,51335,51336,75910,75911,75912'),
 ('christian mccaffrey,garrett wilson,josh allen,travis kelce,rachaad white,george pickens,aaron jones,,,', 126.35999999999991, '14173,14174,14175,14176,14177,14178,14179,14180,14181,14182,14183,14184,38749,38750,38751,38752,38753,38754,38755,38756,38757,38758,38759,38760,63325,63326,63327,63328,63329,63330,63331,63332,63333,63334,63335,63336,87901,87902,87903,87904,87905,87906,87907,87908,87909,87910,87911,87912'),
 ('christian mccaffrey,garrett wilson,josh allen,travis kelce,malik nabers,joe mixon,aaron jones,,,', 126.35000000000008, '14461,14462,14463,14464,14465,14466,14467,14468,14469,14470,14471,14472,39037,39038,39039,39040,39041,39042,39043,39044,39045,39046,39047,39048,63613,63614,63615,63616,63617,63618,63619,63620,63621,63622,63623,63624,88189,88190,88191,88192,88193,88194,88195,88196,88197,88198,88199,88200'),
 ('christian mccaffrey,garrett wilson,josh allen,derrick henry,trey mcbride,george pickens,aaron jones,,,', 126.21000000000002, '13597,13598,13599,13600,13601,13602,13603,13604,13605,13606,13607,13608,38173,38174,38175,38176,38177,38178,38179,38180,38181,38182,38183,38184,62749,62750,62751,62752,62753,62754,62755,62756,62757,62758,62759,62760,87325,87326,87327,87328,87329,87330,87331,87332,87333,87334,87335,87336'),
 ('ceedee lamb,jahmyr gibbs,josh allen,travis kelce,rachaad white,george pickens,aaron jones,,,', 126.18000000000006, '100189,100190,100191,100192,100193,100194,100195,100196,100197,100198,100199,100200,124765,124766,124767,124768,124769,124770,124771,124772,124773,12 ... (37 characters truncated) ... 43,149344,149345,149346,149347,149348,149349,149350,149351,149352,173917,173918,173919,173920,173921,173922,173923,173924,173925,173926,173927,173928'),
 ('ceedee lamb,jahmyr gibbs,josh allen,travis kelce,malik nabers,joe mixon,aaron jones,,,', 126.17000000000006, '100477,100478,100479,100480,100481,100482,100483,100484,100485,100486,100487,100488,125053,125054,125055,125056,125057,125058,125059,125060,125061,12 ... (37 characters truncated) ... 31,149632,149633,149634,149635,149636,149637,149638,149639,149640,174205,174206,174207,174208,174209,174210,174211,174212,174213,174214,174215,174216'),
 ('christian mccaffrey,jahmyr gibbs,sam laporta,stefon diggs,lamar jackson,george pickens,aaron jones,,,', 126.07999999999994, '8605,8606,8607,8608,8609,8610,8611,8612,8613,8614,8615,8616,33181,33182,33183,33184,33185,33186,33187,33188,33189,33190,33191,33192,57757,57758,57759,57760,57761,57762,57763,57764,57765,57766,57767,57768,82333,82334,82335,82336,82337,82338,82339,82340,82341,82342,82343,82344'),
 ('christian mccaffrey,jahmyr gibbs,josh allen,travis kelce,malik nabers,,chris godwin,dandre swift,,', 126.05999999999996, '2152,2153,2154,2248,2249,2250,26728,26729,26730,26824,26825,26826,51304,51305,51306,51400,51401,51402,75880,75881,75882,75976,75977,75978'),
 ('christian mccaffrey,garrett wilson,josh allen,travis kelce,rachaad white,joe mixon,chris godwin,,,', 126.05000000000011, '14101,14102,14103,14104,14105,14106,14107,14108,14109,14110,14111,14112,38677,38678,38679,38680,38681,38682,38683,38684,38685,38686,38687,38688,63253,63254,63255,63256,63257,63258,63259,63260,63261,63262,63263,63264,87829,87830,87831,87832,87833,87834,87835,87836,87837,87838,87839,87840'),
 ('christian mccaffrey,jahmyr gibbs,josh allen,travis kelce,malik nabers,,aaron jones,,jordan addison,', 126.02999999999999, '2175,2178,2181,26751,

In [163]:
%%sql
WITH rosters AS (
SELECT
    team_id,
    GROUP_CONCAT(IIF(team_score > 0, position, "")) AS team,
    SUM(team_score) AS team_total
FROM (
    SELECT *
    FROM image_roster_1combo_eval_2024
    WHERE 1=1
    ORDER BY CAST(round AS INT) ASC
    )
GROUP BY 1
)
SELECT
    team,
    AVG(team_total),
    GROUP_CONCAT(team_id) AS team_list
FROM rosters
GROUP BY 1
ORDER BY 2 DESC
LIMIT 20

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db


Done.


[('RB,RB,QB,WR,TE,WR,RB,,,', 127.22999999999986, '2845,2846,2847,2848,2849,2850,2851,2852,2853,2854,2855,2856,27421,27422,27423,27424,27425,27426,27427,27428,27429,27430,27431,27432,51997,51998,51999,52000,52001,52002,52003,52004,52005,52006,52007,52008,76573,76574,76575,76576,76577,76578,76579,76580,76581,76582,76583,76584'),
 ('RB,RB,QB,TE,WR,,RB,WR,,', 126.36999999999996, '2182,2183,2184,26758,26759,26760,51334,51335,51336,75910,75911,75912'),
 ('RB,WR,QB,TE,RB,WR,RB,,,', 126.35999999999991, '14173,14174,14175,14176,14177,14178,14179,14180,14181,14182,14183,14184,38749,38750,38751,38752,38753,38754,38755,38756,38757,38758,38759,38760,63325,63326,63327,63328,63329,63330,63331,63332,63333,63334,63335,63336,87901,87902,87903,87904,87905,87906,87907,87908,87909,87910,87911,87912'),
 ('RB,WR,QB,TE,WR,RB,RB,,,', 126.35000000000008, '14461,14462,14463,14464,14465,14466,14467,14468,14469,14470,14471,14472,39037,39038,39039,39040,39041,39042,39043,39044,39045,39046,39047,39048,63613,63614,63615,63616,63617,63618,63619,63620,63621,63622,63623,63624,88189,88190,88191,88192,88193,88194,88195,88196,88197,88198,88199,88200'),
 ('RB,WR,QB,RB,TE,WR,RB,,,', 126.21000000000002, '13597,13598,13599,13600,13601,13602,13603,13604,13605,13606,13607,13608,38173,38174,38175,38176,38177,38178,38179,38180,38181,38182,38183,38184,62749,62750,62751,62752,62753,62754,62755,62756,62757,62758,62759,62760,87325,87326,87327,87328,87329,87330,87331,87332,87333,87334,87335,87336'),
 ('WR,RB,QB,TE,RB,WR,RB,,,', 126.18000000000006, '100189,100190,100191,100192,100193,100194,100195,100196,100197,100198,100199,100200,124765,124766,124767,124768,124769,124770,124771,124772,124773,12 ... (37 characters truncated) ... 43,149344,149345,149346,149347,149348,149349,149350,149351,149352,173917,173918,173919,173920,173921,173922,173923,173924,173925,173926,173927,173928'),
 ('WR,RB,QB,TE,WR,RB,RB,,,', 126.17000000000006, '100477,100478,100479,100480,100481,100482,100483,100484,100485,100486,100487,100488,125053,125054,125055,125056,125057,125058,125059,125060,125061,12 ... (37 characters truncated) ... 31,149632,149633,149634,149635,149636,149637,149638,149639,149640,174205,174206,174207,174208,174209,174210,174211,174212,174213,174214,174215,174216'),
 ('RB,RB,TE,WR,QB,WR,RB,,,', 126.07999999999994, '8605,8606,8607,8608,8609,8610,8611,8612,8613,8614,8615,8616,33181,33182,33183,33184,33185,33186,33187,33188,33189,33190,33191,33192,57757,57758,57759,57760,57761,57762,57763,57764,57765,57766,57767,57768,82333,82334,82335,82336,82337,82338,82339,82340,82341,82342,82343,82344'),
 ('RB,RB,QB,TE,WR,,WR,RB,,', 126.05999999999996, '2152,2153,2154,2248,2249,2250,26728,26729,26730,26824,26825,26826,51304,51305,51306,51400,51401,51402,75880,75881,75882,75976,75977,75978'),
 ('RB,WR,QB,TE,RB,RB,WR,,,', 126.05000000000011, '14101,14102,14103,14104,14105,14106,14107,14108,14109,14110,14111,14112,38677,38678,38679,38680,38681,38682,38683,38684,38685,38686,38687,38688,63253,63254,63255,63256,63257,63258,63259,63260,63261,63262,63263,63264,87829,87830,87831,87832,87833,87834,87835,87836,87837,87838,87839,87840'),
 ('RB,RB,QB,TE,WR,,RB,,WR,', 126.02999999999999, '2175,2178,2181,26751,26754,26757,51327,51330,51333,75903,75906,75909'),
 ('WR,RB,QB,RB,TE,WR,RB,,,', 126.02999999999997, '99613,99614,99615,99616,99617,99618,99619,99620,99621,99622,99623,99624,124189,124190,124191,124192,124193,124194,124195,124196,124197,124198,124199, ... (25 characters truncated) ... 67,148768,148769,148770,148771,148772,148773,148774,148775,148776,173341,173342,173343,173344,173345,173346,173347,173348,173349,173350,173351,173352'),
 ('RB,RB,QB,TE,,WR,RB,WR,,', 125.92000000000003, '1894,1895,1896,2086,2087,2088,26470,26471,26472,26662,26663,26664,51046,51047,51048,51238,51239,51240,75622,75623,75624,75814,75815,75816'),
 ('WR,RB,QB,TE,RB,RB,WR,,,', 125.8699999999999, '100117,100118,100119,100120,100121,100122,100123,100124,100125,100126,100127,100128,124693,124694,124695,124696,124697,124698,124699,124700,12

In [165]:
%%sql
WITH rosters AS (
SELECT
    team_id,
    GROUP_CONCAT(IIF(team_score > 0, position, "")) AS team,
    SUM(team_score) AS team_total
FROM (
    SELECT *
    FROM image_roster_1combo_eval_2024
    WHERE 1=1
    ORDER BY CAST(round AS INT) ASC
    )
GROUP BY 1
)
SELECT
    team,
    AVG(team_total),
    GROUP_CONCAT(team_id) AS team_list
FROM rosters
GROUP BY 1
ORDER BY 2 DESC
LIMIT 20

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[(7176,)]

in a 10 round draft there are
1. 2. 3. 4. 5. 6. 7. 8. 9. 10.

early: 1. 2.

middle: 3. 4.

late: 5. 6. 7. 

dart: 8. 9. 10.

In [192]:
%%sql
WITH team AS (
SELECT
    team_id,
    GROUP_CONCAT(IIF(round IN ('1', '2'), round || ':' || position,NULL)) AS early,
    GROUP_CONCAT(IIF(round IN ('3', '4'), round || ':' || position,NULL)) AS middle,
    GROUP_CONCAT(IIF(round IN ('5', '6', '7'), round || ':' || position,NULL)) AS late,
    SUM(team_score) AS ppg
FROM(
    SELECT *
    FROM image_roster_1combo_eval_2024
    WHERE 1=1
        AND team_score > 0
    ORDER BY CAST(round AS INT) ASC
    )
GROUP BY 1
)
SELECT
    early,
    middle,
    CASE
        WHEN early LIKE '%TE%' OR early LIKE '%QB%' THEN 'early onsie'
        WHEN middle LIKE '%TE%' OR middle LIKE '%QB%' THEN 'middle onsie'
        ELSE 'late onsie'
        END AS onsie,
    AVG(ppg) AS avg_score
FROM team
WHERE 1=1
    AND middle LIKE '%,%'
GROUP BY 1, 2, 3
ORDER BY AVG(ppg) DESC

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('1:WR,2:WR', '3:QB,4:WR', 'middle onsie', 120.91956521739145),
 ('1:RB,2:RB', '3:WR,4:TE', 'middle onsie', 120.5661426340447),
 ('1:RB,2:WR', '3:WR,4:WR', 'late onsie', 120.05914596273279),
 ('1:RB,2:WR', '3:TE,4:RB', 'middle onsie', 119.9261426340441),
 ('1:WR,2:RB', '3:WR,4:WR', 'late onsie', 119.87914596273315),
 ('1:RB,2:WR', '3:RB,4:TE', 'middle onsie', 119.85614263404402),
 ('1:WR,2:WR', '3:TE,4:WR', 'middle onsie', 119.7639084507041),
 ('1:WR,2:RB', '3:TE,4:RB', 'middle onsie', 119.74614263404474),
 ('1:WR,2:RB', '3:RB,4:TE', 'middle onsie', 119.67614263404411),
 ('1:WR,2:WR', '3:RB,4:WR', 'late onsie', 119.16914596273284),
 ('1:RB,2:RB', '3:QB,4:TE', 'middle onsie', 117.67884540117417),
 ('1:RB,2:WR', '3:WR,4:RB', 'late onsie', 117.65216033315963),
 ('1:WR,2:RB', '3:WR,4:RB', 'late onsie', 117.47216033316042),
 ('1:RB,2:WR', '3:QB,4:TE', 'middle onsie', 117.03813789868647),
 ('1:WR,2:RB', '3:QB,4:TE', 'middle onsie', 116.85813789868722),
 ('1:WR,2:WR', '3:RB,4:RB', 'late onsie', 116.76216033316067),
 ('1:RB,2:RB', '3:TE,4:QB', 'middle onsie', 116.66884540117466),
 ('1:RB,2:WR', '3:TE,4:QB', 'middle onsie', 116.02813789868691),
 ('1:RB,2:RB', '3:QB,4:WR', 'middle onsie', 115.95303939962497),
 ('1:RB,2:RB', '3:TE,4:WR', 'middle onsie', 115.94696512233267),
 ('1:WR,2:RB', '3:TE,4:QB', 'middle onsie', 115.84813789868649),
 ('1:RB,2:WR', '3:QB,4:WR', 'middle onsie', 115.60383208255202),
 ('1:RB,2:WR', '3:QB,4:RB', 'middle onsie', 115.56435272045069),
 ('1:WR,2:RB', '3:QB,4:WR', 'middle onsie', 115.42383208255245),
 ('1:WR,2:RB', '3:QB,4:RB', 'middle onsie', 115.38435272044953),
 ('1:WR,2:WR', '3:QB,4:RB', 'middle onsie', 115.35080206378899),
 ('1:RB,2:RB', '3:WR,4:QB', 'middle onsie', 115.19435272045045),
 ('1:RB,2:WR', '3:WR,4:QB', 'middle onsie', 115.16080206378946),
 ('1:WR,2:RB', '3:WR,4:QB', 'middle onsie', 114.98080206379021),
 ('1:RB,2:WR', '3:RB,4:QB', 'middle onsie', 114.48435272044942),
 ('1:RB,2:WR', '3:TE,4:WR', 'middle onsie', 114.33489328474865),
 ('1:WR,2:RB', '3:RB,4:QB', 'middle onsie', 114.30435272045084),
 ('1:WR,2:WR', '3:RB,4:QB', 'middle onsie', 114.27080206379006),
 ('1:WR,2:RB', '3:TE,4:WR', 'middle onsie', 114.15489328474786),
 ('1:WR,2:WR', '3:QB,4:TE', 'middle onsie', 114.15210697977793),
 ('1:RB,2:RB', '3:WR,4:WR', 'late onsie', 113.67298282144638),
 ('1:WR,2:WR', '3:TE,4:QB', 'middle onsie', 113.14210697977866),
 ('1:RB,2:WR', '3:WR,4:TE', 'middle onsie', 113.10953669963551),
 ('1:RB,2:WR', '3:RB,4:WR', 'late onsie', 112.96298282144771),
 ('1:WR,2:RB', '3:WR,4:TE', 'middle onsie', 112.92953669963573),
 ('1:WR,2:RB', '3:RB,4:WR', 'late onsie', 112.78298282144769),
 ('1:WR,2:WR', '3:TE,4:RB', 'middle onsie', 112.28953669963437),
 ('1:WR,2:WR', '3:RB,4:TE', 'middle onsie', 112.21953669963469)]

In [183]:
%%sql
WITH team AS (
SELECT
    team_id,
    round,
    position,
    name,
    IIF(round IN ('1', '2'), position,NULL) AS early,
    IIF(round IN ('3', '4'), position,NULL) AS middle,
    team_score
FROM(SELECT * FROM image_roster_1combo_eval_2024 ORDER BY CAST(round AS INT) ASC)
where round IN ('3', '4') AND position != 'QB' AND POSITION != 'TE'
and team_score > 0
)
SELECT
   *
FROM team
LIMIT 10

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[(9225, '3', 'WR', 'deebo samuel', None, 'WR', 15.52),
 (9227, '3', 'WR', 'deebo samuel', None, 'WR', 15.52),
 (9237, '3', 'WR', 'deebo samuel', None, 'WR', 15.52),
 (9239, '3', 'WR', 'deebo samuel', None, 'WR', 15.52),
 (9243, '3', 'WR', 'deebo samuel', None, 'WR', 15.52),
 (9246, '3', 'WR', 'deebo samuel', None, 'WR', 15.52),
 (9249, '3', 'WR', 'deebo samuel', None, 'WR', 15.52),
 (9250, '3', 'WR', 'deebo samuel', None, 'WR', 15.52),
 (9251, '3', 'WR', 'deebo samuel', None, 'WR', 15.52),
 (9252, '3', 'WR', 'deebo samuel', None, 'WR', 15.52)]

In [112]:
%%sql
SELECT *
FROM nfl_adp_hype
LIMIT 20

 * sqlite:////Users/beoconno/Documents/ff/ff_data/raw_data/nfl/src_code/nfl.db
Done.


[('Aaron Burbridge', 473.0, None, 'WR170', 'WR', '170', '2016', 440.0, 3.0, '48', 'aaron burbridge', 422, 'Aaron Burbridge', 'SFO', 'WR', 23, 16, 3, 0, 0, 0, 0, 0, 0, 0, None, 0, 16, 7, 88, 12.57, 0, 1.0, 1, 0, None, None, 7.0, 13.8, 14.8, 10.3, None, 171, None, 'BurbAa00', '2016', 'Aaron Burbridge', 'false', 'false', 0.44, 2.33, 0.86, 10.5, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, None, None, 0.0, 0.0, None, None, 0.438, 0.438, 0.0, 5.5, None, None, None, 'aaron burbridge', 'aaron burbridge', 'Aaron Burbridge', '2016', 'SFO', 'WR', '2016', '2016', 1, None, None, 0, 0, None, None, 183, 188, 165, 169, 8, 1.58, 1.58, -0.72, ''),
 ('Aaron Dobson', 483.0, None, 'WR175', 'WR', '175', '2016', 450.0, 3.0, '49', 'aaron dobson', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'aaron dobson', 'Aaron Dobson', '2016', None, 'WR', None, None, 1, None, None, 0, 0, None, None, 43, 43, 43, 43, 8, 1.37, 1.37, 0, ''),
 (None, None, None, None, None, None, None, None, None, None, None, 617, 'Aaron Fuller', 'SEA', 'WR', 24, 1, 0, 0, 0, 0, 0, 0, 0, 0, None, 0, 0, 0, 0, None, 0, 0.0, 0, 0, None, None, None, None, None, None, None, 245, None, 'FullAa01', '2021', 'Aaron Fuller', 'false', 'false', None, None, None, None, 0.06, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, 0.0, 0.0, None, None, 0.0, None, 0.0, 0.0, None, None, None, 'aaron fuller', 'aaron fuller', 'Aaron Fuller', '2021', 'SEA', 'WR', '2021', '2021', 1, None, None, 0, 0, None, None, 251, 251, 251, 251, None, None, None, 0, ''),
 (None, None, None, None, None, None, None, None, None, None, None, 544, 'Aaron Green', 'LAR', 'RB', 24, 2, 0, 0, 0, 0, 0, 0, 0, 0, None, 0, 0, 0, 0, None, 0, None, 0, 0, None, None, None, None, None, None, None, 161, None, 'GreeAa01', '2016', 'Aaron Green', 'false', 'false', None, None, None, None, 0.13, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, 0.0, 0.0, None, None, 0.0, None, 0.0, 0.0, None, None, None, 'aaron green', 'aaron green', 'Aaron Green', '2016', 'LAR', 'RB', '2016', '2016', 1, None, None, 0, 0, None, None, 158, 158, 158, 155, None, None, None, 0, ''),
 ('Aaron Hernandez', 385.0, None, 'TE44', 'TE', '44', '2017', 344.0, 5.0, '39', 'aaron hernandez', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'aaron hernandez', 'Aaron Hernandez', '2017', None, 'TE', None, None, 1, None, None, 0, 0, None, None, 95, 95, 95, 95, 8, 3.97, 3.97, 0, ''),
 ('Aaron Hernandez', 413.0, None, 'TE58', 'TE', '58', '2018', 400.0, 3.0, '42', 'aaron hernandez', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'aaron hernandez', 'Aaron Hernandez', '2018', None, 'TE', None, None, 2, None, None, 0, 0, None, None, 128, 128, 128, 128, 8, 2.77, 2.77, 0, ''),
 ('Aaron Hernandez', 416.0, None, 'TE46', 'TE', '46', '2019', 418.0, 6.0, '42', 'aaron hernandez', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,

In [133]:
%%sql
PRAGMA table_info(nfl_adp_hype);

 * sqlite:////Users/beoconno/Documents/ff/ff_data/raw_data/nfl/src_code/nfl.db
Done.


[(0, 'adp_player', 'TEXT', 0, None, 0),
 (1, 'adp_ovr', 'REAL', 0, None, 0),
 (2, 'adp_team', 'TEXT', 0, None, 0),
 (3, 'adp_pos', 'TEXT', 0, None, 0),
 (4, 'adp_pos_only', '', 0, None, 0),
 (5, 'adp_pos_rank', '', 0, None, 0),
 (6, 'adp_year', 'TEXT', 0, None, 0),
 (7, 'adp_avg', 'REAL', 0, None, 0),
 (8, 'dr10_pick', '', 0, None, 0),
 (9, 'dr10', '', 0, None, 0),
 (10, 'adp_join_name', '', 0, None, 0),
 (11, 'ranker', 'INT', 0, None, 0),
 (12, 'player', 'TEXT', 0, None, 0),
 (13, 'team', 'TEXT', 0, None, 0),
 (14, 'fantasy_pos', 'TEXT', 0, None, 0),
 (15, 'age', 'INT', 0, None, 0),
 (16, 'g', 'INT', 0, None, 0),
 (17, 'gs', 'INT', 0, None, 0),
 (18, 'pass_cmp', 'INT', 0, None, 0),
 (19, 'pass_att', 'INT', 0, None, 0),
 (20, 'pass_yds', 'INT', 0, None, 0),
 (21, 'pass_td', 'INT', 0, None, 0),
 (22, 'pass_int', 'INT', 0, None, 0),
 (23, 'rush_att', 'INT', 0, None, 0),
 (24, 'rush_yds', 'INT', 0, None, 0),
 (25, 'rush_yds_per_att', 'REAL', 0, None, 0),
 (26, 'rush_td', 'INT', 0, None, 0),
 (27, 'targets', 'INT', 0, None, 0),
 (28, 'rec', 'INT', 0, None, 0),
 (29, 'rec_yds', 'INT', 0, None, 0),
 (30, 'rec_yds_per_rec', 'REAL', 0, None, 0),
 (31, 'rec_td', 'INT', 0, None, 0),
 (32, 'fumbles', 'REAL', 0, None, 0),
 (33, 'fumbles_lost', 'INT', 0, None, 0),
 (34, 'all_td', 'INT', 0, None, 0),
 (35, 'two_pt_md', 'REAL', 0, None, 0),
 (36, 'two_pt_pass', 'REAL', 0, None, 0),
 (37, 'fantasy_points', 'REAL', 0, None, 0),
 (38, 'fantasy_points_ppr', 'REAL', 0, None, 0),
 (39, 'draftkings_points', 'REAL', 0, None, 0),
 (40, 'fanduel_points', 'REAL', 0, None, 0),
 (41, 'vbd', 'REAL', 0, None, 0),
 (42, 'fantasy_rank_pos', 'INT', 0, None, 0),
 (43, 'fantasy_rank_overall', 'REAL', 0, None, 0),
 (44, 'player_id', 'TEXT', 0, None, 0),
 (45, 'Year', 'TEXT', 0, None, 0),
 (46, 'cleaned_player', '', 0, None, 0),
 (47, 'pro_bowl', '', 0, None, 0),
 (48, 'all_pro', '', 0, None, 0),
 (49, 'fppg', '', 0, None, 0),
 (50, 'fppgs', '', 0, None, 0),
 (51, 'fppg_ppr', '', 0, None, 0),
 (52, 'fantasy_points_hppr', '', 0, None, 0),
 (53, 'pgp', '', 0, None, 0),
 (54, 'pass_att_pg', '', 0, None, 0),
 (55, 'rush_att_pg', '', 0, None, 0),
 (56, 'targets_pg', '', 0, None, 0),
 (57, 'pass_cmp_pg', '', 0, None, 0),
 (58, 'pass_yds_pg', '', 0, None, 0),
 (59, 'pass_td_pg', '', 0, None, 0),
 (60, 'pass_td_rate', '', 0, None, 0),
 (61, 'pass_att_ptd', '', 0, None, 0),
 (62, 'rush_td_pg', '', 0, None, 0),
 (63, 'rush_yds_pg', '', 0, None, 0),
 (64, 'rush_yds_per_td', '', 0, None, 0),
 (65, 'rush_att_per_td', '', 0, None, 0),
 (66, 'rec_pg', '', 0, None, 0),
 (67, 'catch_rate', '', 0, None, 0),
 (68, 'rec_td_pg', '', 0, None, 0),
 (69, 'rec_yds_pg', '', 0, None, 0),
 (70, 'tar_per_td', '', 0, None, 0),
 (71, 'rec_per_td', '', 0, None, 0),
 (72, 'rec_yds_per_td', '', 0, None, 0),
 (73, 'nfl_join_name', '', 0, None, 0),
 (74, 'name', '', 0, None, 0),
 (75, 'c_player', '', 0, None, 0),
 (76, 'c_year', '', 0, None, 0),
 (77, 'c_team', '', 0, None, 0),
 (78, 'c_pos', '', 0, None, 0),
 (79, 'fyd', '', 0, None, 0),
 (80, 'lyd', '', 0, None, 0),
 (81, 'yid', '', 0, None, 0),
 (82, 'next_fppg', '', 0, None, 0),
 (83, 'next_fppg_ppr', '', 0, None, 0),
 (84, 'prior_fppg', '', 0, None, 0),
 (85, 'prior_fppg_ppr', '', 0, None, 0),
 (86, 'team_year_prior', '', 0, None, 0),
 (87, 'team_next_year', '', 0, None, 0),
 (88, 'actual_pos_fin_fppg_ppr', '', 0, None, 0),
 (89, 'actual_pos_fin_fppg_std', '', 0, None, 0),
 (90, 'actual_pos_fin_tot_ppr', '', 0, None, 0),
 (91, 'actual_pos_fin_tot_std', '', 0, None, 0),
 (92, 'n_in_exp_fppg_ppr', '', 0, None, 0),
 (93, 'exp_fppg_ppr', '', 0, None, 0),
 (94, 'pick_hype', '', 0, None, 0),
 (95, 'pick_accuracy', '', 0, None, 0),
 (96, 'team_change', '', 0, None, 0)]

In [155]:
%%sql
SELECT
    IIF(c_team LIKE '%TM',team_year_prior,c_team) AS fill_team,
    c_year,
    c_pos,
    COUNT(1) AS players_drafted,
    MIN((IIF(depth_rank_adp=1,adp_ovr,NULL))) AS adp_ovr_depth1,
MIN((IIF(depth_rank_adp=2,adp_ovr,NULL))) AS adp_ovr_depth2,
MIN((IIF(depth_rank_adp=3,adp_ovr,NULL))) AS adp_ovr_depth3,

MIN((IIF(depth_rank_adp=1,age,NULL))) AS age_depth1,
MIN((IIF(depth_rank_adp=2,age,NULL))) AS age_depth2,
MIN((IIF(depth_rank_adp=3,age,NULL))) AS age_depth3,

MIN((IIF(depth_rank_adp=1,g,NULL))) AS g_depth1,
MIN((IIF(depth_rank_adp=2,g,NULL))) AS g_depth2,
MIN((IIF(depth_rank_adp=3,g,NULL))) AS g_depth3,

MIN((IIF(depth_rank_adp=1,adp_pos_rank,NULL))) AS adp_pos_rank_d1,
MIN((IIF(depth_rank_adp=2,adp_pos_rank,NULL))) AS adp_pos_rank_d2,
MIN((IIF(depth_rank_adp=3,adp_pos_rank,NULL))) AS adp_pos_rank_d3
FROM nfl_adp_hype
WHERE
    c_team IS NOT NULL
    AND c_pos = 'RB'
    AND c_team = 'LAR'
GROUP BY 1, 2, 3
ORDER BY 2 DESC, 1 ASC
LIMIT 30

 * sqlite:////Users/beoconno/Documents/ff/ff_data/raw_data/nfl/src_code/nfl.db
Done.


[('LAR', '2024', 'RB', 2, 20.0, 115.0, None, None, None, None, None, None, None, 7, 36, None),
 ('LAR', '2023', 'RB', 5, 257.0, 274.0, None, 23, 22, None, 12, 10, None, 71, 73, None),
 ('LAR', '2022', 'RB', 5, 43.0, 116.0, None, 23, 25, None, 15, 10, None, 20, 40, None),
 ('LAR', '2021', 'RB', 5, 45.0, 97.0, None, 24, 26, None, 12, 17, None, 22, 39, None),
 ('LAR', '2020', 'RB', 5, 60.0, 124.0, 245.0, 21, 23, 27, 13, 15, 16, 25, 45, 73),
 ('LAR', '2019', 'RB', 4, 14.0, 120.0, 190.0, 25, 22, 26, 15, 13, 14, 9, 43, 60),
 ('LAR', '2018', 'RB', 4, 1.0, 217.0, 297.0, 24, 22, 25, 14, 4, 12, 1, 67, 80),
 ('LAR', '2017', 'RB', 4, 18.0, 395.0, None, 23, 24, None, 15, 11, None, 10, 114, None),
 ('LAR', '2016', 'RB', 6, 4.0, 256.0, None, 22, 26, None, 16, 11, None, 1, 80, None)]

In [153]:
%%sql
SELECT
    IIF(c_team LIKE '%TM',team_year_prior,c_team) AS fill_team,
    c_year,
    c_pos,
    name,
    adp_ovr,
    fantasy_points_ppr,
    depth_rank_tot,
    fppg_ppr,
    depth_rank_ppg_ppr,
    depth_rank_adp
FROM nfl_adp_hype
WHERE 1=1
    AND c_pos = 'RB'
    AND C_team = 'LAR'
ORDER BY fppg_ppr DESC 
LIMIT 10

 * sqlite:////Users/beoconno/Documents/ff/ff_data/raw_data/nfl/src_code/nfl.db
Done.


[('LAR', '2018', 'RB', 'todd gurley', 1.0, 372.1, 1, 26.58, 1, 1),
 ('LAR', '2017', 'RB', 'todd gurley', 18.0, 383.3, 1, 25.55, 1, 1),
 ('LAR', '2023', 'RB', 'kyren williams', 257.0, 255.0, 1, 21.25, 1, 1),
 ('LAR', '2019', 'RB', 'todd gurley', 14.0, 219.4, 1, 14.63, 1, 1),
 ('LAR', '2021', 'RB', 'darrell henderson', 45.0, 163.4, 1, 13.62, 1, 1),
 ('LAR', '2016', 'RB', 'todd gurley', 4.0, 198.2, 1, 12.39, 1, 1),
 ('LAR', '2023', 'RB', 'darrell henderson', None, 43.5, 3, 10.88, 2, None),
 ('LAR', '2022', 'RB', 'cam akers', 43.0, 141.3, 1, 9.42, 1, 1),
 ('LAR', '2020', 'RB', 'darrell henderson', 124.0, 130.3, 1, 8.69, 1, 2),
 ('LAR', '2021', 'RB', 'sony michel', 97.0, 146.3, 2, 8.61, 2, 2)]

In [166]:
%%sql
SELECT * 
FROM nfl_window_stats
limit 10

 * sqlite:////Users/beoconno/Documents/ff/ff_data/raw_data/nfl/src_code/nfl.db
Done.


[('Aaron Burbridge', 473.0, None, 'WR170', 'WR', 170, '2016', 440.0, 3.0, '48', 'aaron burbridge', 422, 'Aaron Burbridge', 'SFO', 'WR', 23, 16, 3, 0, 0, 0, 0, 0, 0, 0, None, 0, 16, 7, 88, 12.57, 0, 1.0, 1, 0, None, None, 7.0, 13.8, 14.8, 10.3, None, 171, None, 'BurbAa00', '2016', 'Aaron Burbridge', 'false', 'false', 0.44, 2.33, 0.86, 10.5, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, None, None, 0.0, 0.0, None, None, 0.438, 0.438, 0.0, 5.5, None, None, None, 'aaron burbridge', 'aaron burbridge', 'Aaron Burbridge', '2016', 'SFO', 'WR', '2016', '2016', 1, None, None, 0, 0, None, None, 'SFO', 183, 188, 165, 169),
 ('Aaron Dobson', 483.0, None, 'WR175', 'WR', 175, '2016', 450.0, 3.0, '49', 'aaron dobson', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'aaron dobson', 'Aaron Dobson', '2016', None, 'WR', None, None, 1, None, None, 0, 0, None, None, None, 43, 43, 43, 43),
 (None, None, None, None, None, None, None, None, None, None, None, 617, 'Aaron Fuller', 'SEA', 'WR', 24, 1, 0, 0, 0, 0, 0, 0, 0, 0, None, 0, 0, 0, 0, None, 0, 0.0, 0, 0, None, None, None, None, None, None, None, 245, None, 'FullAa01', '2021', 'Aaron Fuller', 'false', 'false', None, None, None, None, 0.06, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, 0.0, 0.0, None, None, 0.0, None, 0.0, 0.0, None, None, None, 'aaron fuller', 'aaron fuller', 'Aaron Fuller', '2021', 'SEA', 'WR', '2021', '2021', 1, None, None, 0, 0, None, None, 'SEA', 251, 251, 251, 251),
 (None, None, None, None, None, None, None, None, None, None, None, 544, 'Aaron Green', 'LAR', 'RB', 24, 2, 0, 0, 0, 0, 0, 0, 0, 0, None, 0, 0, 0, 0, None, 0, None, 0, 0, None, None, None, None, None, None, None, 161, None, 'GreeAa01', '2016', 'Aaron Green', 'false', 'false', None, None, None, None, 0.13, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, 0.0, 0.0, None, None, 0.0, None, 0.0, 0.0, None, None, None, 'aaron green', 'aaron green', 'Aaron Green', '2016', 'LAR', 'RB', '2016', '2016', 1, None, None, 0, 0, None, None, 'LAR', 158, 158, 158, 155),
 ('Aaron Hernandez', 385.0, None, 'TE44', 'TE', 44, '2017', 344.0, 5.0, '39', 'aaron hernandez', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'aaron hernandez', 'Aaron Hernandez', '2017', None, 'TE', None, None, 1, None, None, 0, 0, None, None, None, 95, 95, 95, 95),
 ('Aaron Hernandez', 413.0, None, 'TE58', 'TE', 58, '2018', 400.0, 3.0, '42', 'aaron hernandez', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'aaron hernandez', 'Aaron Hernandez', '2018', None, 'TE', None, None, 2, None, None, 0, 0, None, None, None, 128, 128, 128, 128),
 ('Aaron Hernandez', 416.0, None, 'TE46', 'TE', 46, '2019', 418.0, 6.0, '42', 'aaron hernandez', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'aaron her

In [175]:
%%sql
SELECT * 
FROM nfl_position_groups
WHERE TEAM IS NOT NULL AND TEAM != '2TM' AND TEAM = 'ARI' AND c_pos != 'TE'
order by c_pos,c_year DESC
limit 30

 * sqlite:////Users/beoconno/Documents/ff/ff_data/raw_data/nfl/src_code/nfl.db
Done.


[('ARI', '2018', 'FB', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 0, 0, 0, 0, 0, 2, 2, 0, 7, 7, 37, 0, 0.0, 0, 0, 10.9, 1.09, None, None, 0, 1),
 ('ARI', '2021', 'PK', 178.0, None, None, None, None, None, None, None, None, 9, None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 0, 1),
 ('ARI', '2024', 'QB', 71.0, None, None, None, None, None, None, None, None, 8, None, None, 6, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 19.28, 0.98, 0, 1),
 ('ARI', '2023', 'QB', 203.0, 273.0, None, 26, 24, None, 8, 7, None, 29, 36, None, 5, 1, None, 188, 289, 1861, 10, 7, 52, 274, 4, 0, 0, 0, 0, 9.0, 1, 4, 153.9, 19.37, 24.82, 6.59, -5.449999999999999, 2),
 ('ARI', '2022', 'QB', 55.0, None, None, 25, None, None, 11, None, None, 5, None, None, 4, None, None, 432, 663, 3962, 17, 17, 100, 520, 3, 0, 0, 0, 0, 13.0, 5, 3, 260.5, 37.160000000000004, 20.84, -0.62, -2.61, 4),
 ('ARI', '2021', 'QB', 42.0, None, None, 24, None, None, 14, None, None, 3, None, None, 3, None, None, 413, 589, 4563, 27, 11, 113, 466, 5, 0, 0, 7, 0, 18.0, 1, 5, 343.8, 27.62, 22.24, -1.43, -0.78, 3),
 ('ARI', '2020', 'QB', 57.0, None, None, 23, None, None, 16, None, None, 5, None, None, 2, None, None, 386, 574, 4076, 27, 13, 137, 834, 11, 0, 0, 0, 0, 10.0, 4, 11, 386.4, 25.21, 20.84, 3.01, 2.83, 2),
 ('ARI', '2019', 'QB', 97.0, None, None, 22, None, None, 16, None, None, 11, None, None, 1, None, None, 354, 553, 3771, 20, 12, 100, 585, 4, 0, 0, 0, 0, 5.0, 2, 4, 291.40000000000003, 19.86, 18.54, 18.54, -0.71, 2),
 ('ARI', '2018', 'QB', 256.0, 313.0, None, 31, 21, None, 3, 14, None, 31, 37, None, 3, 1, None, 282, 494, 2852, 14, 18, 25, 145, 0, 0, 0, 0, 0, 13.0, 7, 0, 136.60000000000002, 17.79, 24.2, 10.7, -11.91, 3),
 ('ARI', '2017', 'QB', 136.0, None, None, 38, None, None, 7, None, None, 18, None, None, 2, None, None, 338, 597, 3958, 21, 18, 45, 101, 0, 0, 0, 0, 0, 12.0, 2, 0, 214.4, 37.03, 16.31, 0.1, -1.7, 3),
 ('ARI', '2016', 'QB', 79.0, 509.0, None, 37, 32, None, 15, 5, None, 8, 67, None, 1, 1, None, 383, 645, 4425, 28, 17, 17, 35, 0, 0, 0, 0, 0, 14.0, 4, 0, 252.5, 18.09, 20.1, 20.1, -2.01, 2),
 ('ARI', '2024', 'RB', 81.0, 112.0, 302.0, None, None, None, None, None, None, 24, 35, 85, 8, 1, 2, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 26.63, 5.890000000000001, 0, 3),
 ('ARI', '2023', 'RB', 59.0, 277.0, None, 28, 24, None, 13, 8, None, 23, 74, None, 7, 2, None, 0, 0, 0, 0, 0, 312, 1441, 9, 69, 54, 322, 2, 0.0, 0, 11, 296.3, 24.990000000000002, 17.64, 0.73, -0.3900000000000001, 5),
 ('ARI', '2022', 'RB', 28.0, 72.0, None, 23, 26, None, 13, 6, None, 16, 27, None, 3, 5, None, 0, 0, 0, 0, 0, 396, 1582, 14, 140, 103, 746, 2, 3.0, 1, 16, 431.8, 36.69, 38.379999999999995, -5.38, -7.78, 8),
 ('ARI', '2021', 'RB', 62.0, 101.0, None, 25, 26, None, 12, 15, None, 28, 40, None, 4, 5, None, 0, 0, 0, 0, 0, 361, 1495, 18, 103, 89, 762, 3, 3.0, 1, 21, 438.7, 32.98, 21.369999999999997, -1.7899999999999998, 7.75, 5),
 ('ARI', '2020', 'RB', 15.0, 143.0, None, 26, 24, None, 15, 16, None, 11, 48, None, 5, 3, None, 0, 0, 0, 0, 0, 336, 1403, 11, 99, 79, 550, 5, 5.0, 1, 16, 368.3, 23.89, 23.86, 2.19, -0.5500000000000003, 3),
 ('ARI', '2019', 'RB', 5.0, 232.0, 250.0, 28, 23, 31, 13, 13, 1, 5, 70, 79, 4, 2, 4, 0, 0, 0, 0, 0, 155, 652, 6, 68, 48, 475, 5, 1.0, 1, 11, 224.70000000000002, 17.65, 29.13, 4.1000000000000005, -11.48, 3),
 ('ARI', '2018', 'RB', 3.0, 287.0, None, 27, 22, None, 16, 16, None, 3, 79, None, 3, 1, None, 0, 0, 0, 0, 0, 320, 1151, 9, 102, 72, 566, 3, 4.0, 2, 12, 313.7, 19.6, 25.459999999999997, 12.46, -6.11, 3),
 ('ARI', '2017', 'RB', 1.0, 253.0, 301.0, 26, 26, 32, 1, 16, 4, 1, 80, 92, 2, 2, 2, 0, 0, 0, 0, 0, 228, 740, 4, 95, 64, 610, 0, 2.0, 1, 4, 221.0, 31.61, 33.54, 25.67, -11.120000000000001, 6),
 ('ARI', '2016', 'RB', 7.

In [164]:
%%sql
SELECT
    fill_team AS team,
    c_year,
    c_pos,
    -- OVERALL ADP FOR THE PLAYER IN DEPTH 1, 2, 3 OF EACH POSITION GROUP FOR THAT TEAM
    MIN((IIF(depth_rank_adp=1,adp_ovr,NULL))) AS adp_ovr_depth1,
    MIN((IIF(depth_rank_adp=2,adp_ovr,NULL))) AS adp_ovr_depth2,
    MIN((IIF(depth_rank_adp=3,adp_ovr,NULL))) AS adp_ovr_depth3,
    -- AGE OF THE PLAYER FOR THE ADP OF 1, 2, 3 ON EACH POSITION GROUP
    MIN((IIF(depth_rank_adp=1,age,NULL))) AS age_depth1,
    MIN((IIF(depth_rank_adp=2,age,NULL))) AS age_depth2,
    MIN((IIF(depth_rank_adp=3,age,NULL))) AS age_depth3,
    -- GAMES PLAYED FOR THE ADP OF 1, 2, 3 ON EACH TEAMS POSITION GROUP
    MIN((IIF(depth_rank_adp=1,g,NULL))) AS g_depth1,
    MIN((IIF(depth_rank_adp=2,g,NULL))) AS g_depth2,
    MIN((IIF(depth_rank_adp=3,g,NULL))) AS g_depth3,
    -- ADP OF THE PLAYERS POSITION GROUP FOR THE 1, 2, 3RD RANKED DEPTH ON EACH TEAM
    MIN((IIF(depth_rank_adp=1,adp_pos_rank,NULL))) AS adp_pos_rank_d1,
    MIN((IIF(depth_rank_adp=2,adp_pos_rank,NULL))) AS adp_pos_rank_d2,
    MIN((IIF(depth_rank_adp=3,adp_pos_rank,NULL))) AS adp_pos_rank_d3,
    -- THE SEASON OF THE PLAYER IN ADP RANK 1, 2, 3
    MIN((IIF(depth_rank_adp=1,yid,NULL))) AS yid_d1,
    MIN((IIF(depth_rank_adp=2,yid,NULL))) AS yid_d2,
    MIN((IIF(depth_rank_adp=3,yid,NULL))) AS yid_d3,
    -- QB STATS
    SUM(pass_cmp) AS tot_pass_cmp,
    SUM(pass_att) AS tot_pass_att,
    SUM(pass_yds) AS tot_pass_yds,
    SUM(pass_td) AS tot_pass_td,
    SUM(pass_int) AS tot_pass_int,
    -- RB STATS
    SUM(rush_att) AS tot_rush_att,
    SUM(rush_yds) AS tot_rush_yds,
    SUM(rush_td) AS tot_rush_td,
    -- WR STATS
    SUM(targets) AS tot_targets,
    SUM(rec) AS tot_rec,
    SUM(rec_yds) AS tot_rec_yds,
    SUM(rec_td) AS tot_rec_td,
    -- OTHER 
    SUM(fumbles) AS tot_fumbles,
    SUM(fumbles_lost) AS tot_fumbles_lost,
    SUM(all_td) AS tot_all_td,
    -- POINTS
    SUM(fantasy_points_ppr) AS tot_fantasy_points_ppr,
    SUM(fppg_ppr) AS tot_fppg_ppr,
    SUM(exp_fppg_ppr) AS tot_exp_fppg_ppr,
    SUM(pick_hype) AS tot_pick_hype,
    SUM(pick_accuracy) AS tot_pick_accuracy,
    -- OTHER
    -- NUMBER OF PLAYERS TRADED
    SUM(IIF(team_change IS NOT NULL,1,0)) AS cnt_team_change
FROM nfl_adp_hype

 * sqlite:////Users/beoconno/Documents/ff/ff_data/raw_data/nfl/src_code/nfl.db
(sqlite3.OperationalError) no such column: fill_team
[SQL: SELECT
    fill_team AS team,
    c_year,
    c_pos,
    -- OVERALL ADP FOR THE PLAYER IN DEPTH 1, 2, 3 OF EACH POSITION GROUP FOR THAT TEAM
    MIN((IIF(depth_rank_adp=1,adp_ovr,NULL))) AS adp_ovr_depth1,
    MIN((IIF(depth_rank_adp=2,adp_ovr,NULL))) AS adp_ovr_depth2,
    MIN((IIF(depth_rank_adp=3,adp_ovr,NULL))) AS adp_ovr_depth3,
    -- AGE OF THE PLAYER FOR THE ADP OF 1, 2, 3 ON EACH POSITION GROUP
    MIN((IIF(depth_rank_adp=1,age,NULL))) AS age_depth1,
    MIN((IIF(depth_rank_adp=2,age,NULL))) AS age_depth2,
    MIN((IIF(depth_rank_adp=3,age,NULL))) AS age_depth3,
    -- GAMES PLAYED FOR THE ADP OF 1, 2, 3 ON EACH TEAMS POSITION GROUP
    MIN((IIF(depth_rank_adp=1,g,NULL))) AS g_depth1,
    MIN((IIF(depth_rank_adp=2,g,NULL))) AS g_depth2,
    MIN((IIF(depth_rank_adp=3,g,NULL))) AS g_depth3,
    -- ADP OF THE PLAYERS POSITION GROUP FOR THE 1, 2

In [212]:
%%sql
SELECT
    name,
    games,
    fppg_ppr,
    pos_rank,
    rank_on_team,
    team_score,
    17 - games AS game_to_fill,
    SUM(games) OVER(PARTITION BY year, position, team_id ORDER BY rank_on_team ASC) AS game_tot,
    SUM(games) OVER(PARTITION BY year, position, team_id ORDER BY rank_on_team ASC)
    >= 2 * 17 AS fill_games
FROM image_roster_1combo_eval
where 1=1
    --and name = 'james conner'
    and team_id = 6470
    and position = 'RB'
order by rb_cnt desc 
limit 10

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('christian mccaffrey', 16, 24.46, 1, 1, 24.46, 1, 16, 0),
 ('travis etienne', 17, 16.61, 2, 4, 16.61, 0, 33, 0),
 ('jahmyr gibbs', 15, 16.14, 3, 5, 16.14, 2, 48, 1),
 ('james conner', 13, 15.5, 4, 6, 0, 4, 61, 1),
 ('derrick henry', 17, 14.51, 5, 7, 0, 0, 78, 1),
 ('khalil herbert', 12, 9.38, 6, 9, 0, 5, 90, 1)]

In [266]:
%%sql
SELECT *
FROM image_roster_1combo_prep_2024_s7
ORDER BY round ASC
LIMIT 20

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('2024', 'RB', 'christian mccaffrey', '1', 26.09),
 ('2024', 'WR', 'ceedee lamb', '1', 23.2),
 ('2024', 'QB', 'brock purdy', '10', 18.54),
 ('2024', 'RB', 'tony pollard', '10', 11.32),
 ('2024', 'TE', 'david njoku', '10', 10.51),
 ('2024', 'WR', 'ladd mcconkey', '10', 11.22),
 ('2024', 'RB', 'jahmyr gibbs', '2', 19.94),
 ('2024', 'WR', 'garrett wilson', '2', 17.23),
 ('2024', 'QB', 'josh allen', '3', 25.09),
 ('2024', 'RB', 'devon achane', '3', 17.52),
 ('2024', 'TE', 'sam laporta', '3', 17.05),
 ('2024', 'WR', 'deebo samuel', '3', 15.52),
 ('2024', 'QB', 'jalen hurts', '4', 22.62),
 ('2024', 'RB', 'derrick henry', '4', 16.13),
 ('2024', 'TE', 'travis kelce', '4', 15.59),
 ('2024', 'WR', 'stefon diggs', '4', 14.44),
 ('2024', 'QB', 'lamar jackson', '5', 21.29),
 ('2024', 'RB', 'rachaad white', '5', 15.09),
 ('2024', 'TE', 'trey mcbride', '5', 14.4),
 ('2024', 'WR', 'malik nabers', '5', 13.77)]

In [268]:
%%sql
SELECT
    c_year AS year,
    c_pos AS position,
    name,
    dr10_adj AS round,
    exp_fppg_ppr AS fppg_ppr
FROM (
    SELECT
        *,
        -- It is never optimal to select anything but the best player for that position in that round
        -- This dramaticaly reduces the number of possibilities that we have to consider
        ROW_NUMBER() OVER(PARTITION BY c_year, c_pos, dr10_adj ORDER BY exp_fppg_ppr DESC) AS best_in_rnd
    FROM (
        SELECT
            *,
            CASE
                -- Adjust the round for draft slot 7
                WHEN adp_ovr < 4 THEN 0
                -- Even round adjustment
                WHEN (CAST(dr10 AS INT) % 2 != 1) AND dr10_pick <= 2 THEN dr10 - 1
                -- Odd round adjustment
                WHEN (CAST(dr10 AS INT) % 2 = 1) AND dr10_pick <= 5 AND dr10 != '1' THEN dr10 - 1
                ELSE CAST(dr10 AS INT)
                END AS dr10_adj
        FROM nfl_adp_hype
        )
    WHERE 1=1
        AND dr10_adj != 0
    )
WHERE 1=1
    AND c_pos IN ("QB", "WR", "RB", "TE")
    AND dr10 IS NOT NULL
    AND CAST(dr10 AS INT) <= 10
    AND c_year = '2024'
    AND best_in_rnd = 1
ORDER BY dr10_adj
LIMIT 10
;

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('2024', 'RB', 'breece hall', 1, 22.83),
 ('2024', 'WR', 'jamarr chase', 1, 20.03),
 ('2024', 'QB', 'josh allen', 2, 25.09),
 ('2024', 'RB', 'jonathan taylor', 2, 19.29),
 ('2024', 'TE', 'sam laporta', 2, 17.05),
 ('2024', 'WR', 'davante adams', 2, 16.82),
 ('2024', 'QB', 'jalen hurts', 3, 22.62),
 ('2024', 'RB', 'travis etienne', 3, 16.56),
 ('2024', 'WR', 'michael pittman', 3, 15.23),
 ('2024', 'QB', 'patrick mahomes', 4, 22.24)]

In [260]:
%%sql
WITH test AS (
SELECT
    name,
    adp_ovr,
    adp_pos_only,
    adp_pos_rank,
    dr10_pick,
    dr10,
    CASE
        -- Pick specs for odd rounds
        WHEN (CAST(dr10 AS INT) % 2 = 1) AND dr10_pick = 6 THEN 'Rd Start'
        WHEN (CAST(dr10 AS INT) % 2 = 1) AND dr10_pick = 7 THEN 'Pick'
        WHEN (CAST(dr10 AS INT) % 2 != 1) AND dr10_pick = 1 THEN 'Rd End'
        -- Pick specs for even rounds
        WHEN (CAST(dr10 AS INT) % 2 != 1) AND dr10_pick = 4 THEN 'Rd Start'
        WHEN (CAST(dr10 AS INT) % 2 != 1) AND dr10_pick = 4 THEN 'Pick'
        WHEN (CAST(dr10 AS INT) % 2 = 1) AND dr10_pick = 5 AND dr10 != '1' THEN 'Rd End'
        ELSE ''
        END AS pick,
    CASE
        WHEN adp_ovr < 4 THEN 0
        -- Even round adjustment
        WHEN (CAST(dr10 AS INT) % 2 != 1) AND dr10_pick <= 3 THEN dr10 - 1
        -- Odd round adjustment
        WHEN (CAST(dr10 AS INT) % 2 = 1) AND dr10_pick <= 3 AND dr10 != '1' THEN dr10 - 1
        ELSE CAST(dr10 AS INT)
        END AS dr10_adj
FROM nfl_adp_hype
WHERE 1=1
    AND c_year = '2024'
    AND dr10 IN ('1', '2', '3', '4', '5')
ORDER BY adp_ovr ASC
)
SELECT
    dr10_adj,
    count(1) AS row_cnt,
    MIN(adp_ovr) AS adp_start,
    MAX(adp_ovr) AS adp_end,
    MIN(dr10_pick) AS dr10_pick_start,
    MAX(dr10_pick) AS dr10_pick_end
FROM test
GROUP BY 1

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[(0, 3, 1.0, 3.0, 1.0, 3.0),
 (1, 10, 4.0, 13.0, 1.0, 10.0),
 (2, 10, 14.0, 23.0, 1.0, 10.0),
 (3, 10, 24.0, 33.0, 1.0, 10.0),
 (4, 10, 34.0, 43.0, 1.0, 10.0),
 (5, 7, 44.0, 50.0, 4.0, 10.0)]

In [264]:
%%sql
WITH test AS (
SELECT
    name,
    adp_ovr,
    adp_pos_only,
    adp_pos_rank,
    dr10_pick,
    dr10,
    CASE
        WHEN (CAST(dr10 AS INT) % 2 = 1) AND dr10_pick = 7 THEN 'Pick' --Odd round
        WHEN (CAST(dr10 AS INT) % 2 != 1) AND dr10_pick = 4 THEN 'Pick' --Even round
        ELSE ''
        END AS pick,
    CASE
        WHEN adp_ovr < 4 THEN 0
        -- Even round adjustment
        WHEN (CAST(dr10 AS INT) % 2 != 1) AND dr10_pick <= 2 THEN dr10 - 1
        -- Odd round adjustment
        WHEN (CAST(dr10 AS INT) % 2 = 1) AND dr10_pick <= 5 AND dr10 != '1' THEN dr10 - 1
        ELSE CAST(dr10 AS INT)
        END AS dr10_adj
FROM nfl_adp_hype
WHERE 1=1
    AND c_year = '2024'
    AND dr10 IN ('1', '2', '3', '4', '5')
ORDER BY adp_ovr ASC
)
SELECT
    *
FROM test


 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('christian mccaffrey', 1.0, 'RB', 1, 1.0, '1', '', 0),
 ('ceedee lamb', 2.0, 'WR', 1, 2.0, '1', '', 0),
 ('tyreek hill', 3.0, 'WR', 2, 3.0, '1', '', 0),
 ('jamarr chase', 4.0, 'WR', 3, 4.0, '1', '', 1),
 ('justin jefferson', 5.0, 'WR', 4, 5.0, '1', '', 1),
 ('breece hall', 6.0, 'RB', 2, 6.0, '1', '', 1),
 ('bijan robinson', 7.0, 'RB', 3, 7.0, '1', 'Pick', 1),
 ('amon-ra st brown', 8.0, 'WR', 5, 8.0, '1', '', 1),
 ('aj brown', 9.0, 'WR', 6, 9.0, '1', '', 1),
 ('puka nacua', 10.0, 'WR', 7, 10.0, '1', '', 1),
 ('garrett wilson', 11.0, 'WR', 8, 1.0, '2', '', 1),
 ('jahmyr gibbs', 12.0, 'RB', 4, 2.0, '2', '', 1),
 ('jonathan taylor', 13.0, 'RB', 5, 3.0, '2', '', 2),
 ('davante adams', 14.0, 'WR', 9, 4.0, '2', 'Pick', 2),
 ('saquon barkley', 15.0, 'RB', 6, 5.0, '2', '', 2),
 ('marvin harrison', 16.0, 'WR', 10, 6.0, '2', '', 2),
 ('chris olave', 17.0, 'WR', 11, 7.0, '2', '', 2),
 ('drake london', 18.0, 'WR', 12, 8.0, '2', '', 2),
 ('nico collins', 19.0, 'WR', 13, 9.0, '2', '', 2),
 ('kyren williams', 20.0, 'RB', 7, 10.0, '2', '', 2),
 ('devon achane', 21.0, 'RB', 8, 1.0, '3', '', 2),
 ('josh allen', 22.0, 'QB', 1, 2.0, '3', '', 2),
 ('deebo samuel', 23.0, 'WR', 14, 3.0, '3', '', 2),
 ('brandon aiyuk', 24.0, 'WR', 15, 4.0, '3', '', 2),
 ('sam laporta', 25.0, 'TE', 1, 5.0, '3', '', 2),
 ('travis etienne', 26.0, 'RB', 9, 6.0, '3', '', 3),
 ('michael pittman', 27.0, 'WR', 16, 7.0, '3', 'Pick', 3),
 ('jaylen waddle', 28.0, 'WR', 17, 8.0, '3', '', 3),
 ('mike evans', 29.0, 'WR', 18, 9.0, '3', '', 3),
 ('dj moore', 30.0, 'WR', 19, 10.0, '3', '', 3),
 ('derrick henry', 31.0, 'RB', 10, 1.0, '4', '', 3),
 ('jalen hurts', 32.0, 'QB', 2, 2.0, '4', '', 3),
 ('travis kelce', 33.0, 'TE', 2, 3.0, '4', '', 4),
 ('stefon diggs', 34.0, 'WR', 20, 4.0, '4', 'Pick', 4),
 ('cooper kupp', 35.0, 'WR', 21, 5.0, '4', '', 4),
 ('dk metcalf', 36.0, 'WR', 22, 6.0, '4', '', 4),
 ('devonta smith', 37.0, 'WR', 23, 7.0, '4', '', 4),
 ('isiah pacheco', 38.0, 'RB', 11, 8.0, '4', '', 4),
 ('josh jacobs', 39.0, 'RB', 12, 9.0, '4', '', 4),
 ('patrick mahomes', 40.0, 'QB', 3, 10.0, '4', '', 4),
 ('lamar jackson', 41.0, 'QB', 4, 1.0, '5', '', 4),
 ('rachaad white', 42.0, 'RB', 13, 2.0, '5', '', 4),
 ('malik nabers', 43.0, 'WR', 24, 3.0, '5', '', 4),
 ('trey mcbride', 44.0, 'TE', 3, 4.0, '5', '', 4),
 ('james cook', 45.0, 'RB', 14, 5.0, '5', '', 4),
 ('zay flowers', 46.0, 'WR', 25, 6.0, '5', '', 5),
 ('amari cooper', 47.0, 'WR', 26, 7.0, '5', 'Pick', 5),
 ('mark andrews', 48.0, 'TE', 4, 8.0, '5', '', 5),
 ('keenan allen', 49.0, 'WR', 27, 9.0, '5', '', 5),
 ('anthony richardson', 50.0, 'QB', 5, 10.0, '5', '', 5)]

In [282]:
%%sql
SELECT
    c_year AS year,
    c_pos AS position,
    name,
    g AS games,
    dr10 AS round,
    fppg_ppr,
    fantasy_points_ppr AS ppr_total
FROM nfl_adp_hype
WHERE 1=1
    AND c_pos IN ("QB", "WR", "RB", "TE")
    AND fppg_ppr IS NOT NULL
    AND dr10 IS NOT NULL
    AND CAST(dr10 AS INT) <= 15
    AND c_year = '2023'
LIMIT 10

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('2023', 'QB', 'lamar jackson', 16, '4', 20.7, 331.2),
 ('2023', 'QB', 'josh allen', 17, '3', 23.09, 392.6),
 ('2023', 'QB', 'justin fields', 13, '5', 17.71, 230.2),
 ('2023', 'QB', 'joe burrow', 10, '4', 14.72, 147.2),
 ('2023', 'QB', 'deshaun watson', 6, '9', 14.47, 86.8),
 ('2023', 'QB', 'dak prescott', 17, '9', 20.16, 342.8),
 ('2023', 'QB', 'jared goff', 17, '14', 17.01, 289.1),
 ('2023', 'QB', 'anthony richardson', 4, '12', 18.18, 72.7),
 ('2023', 'QB', 'trevor lawrence', 16, '6', 16.41, 262.5),
 ('2023', 'QB', 'patrick mahomes', 16, '2', 17.51, 280.2)]

In [286]:
%%sql
SELECT
    team_id,
    SUM(IIF(position = 'RB', 1,0)) AS rb_cnt
FROM image_roster_1combo_eval
GROUP BY 1
HAVING SUM(IIF(position = 'RB', 1,0)) > 2
ORDER BY 2 desc 

LIMIT 10 

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[(159481, 6),
 (159458, 6),
 (159386, 6),
 (159098, 6),
 (158425, 6),
 (158402, 6),
 (158353, 6),
 (158337, 6),
 (158314, 6),
 (158307, 6)]

In [302]:
%%sql 
WITH cume AS (
    SELECT
        *,
        SUM(games) OVER(PARTITION BY team_id, year, position ORDER BY fppg_ppr DESC) AS game_cume
    FROM image_roster_1combo_eval
    WHERE 1=1
        AND team_id = '159458'
        AND position = 'RB'
    )
SELECT
    *,
    game_cume > (17*2),
    IIF(game_cume > (17*2),games - (game_cume - (17*2)),games) AS games_used,
    IIF(
        IIF(game_cume > (17*2),games - (game_cume - (17*2)),games) <= games,
        games - IIF(game_cume > (17*2),games - (game_cume - (17*2)),games),
        games)AS flexable_games
FROM cume
ORDER BY fppg_ppr DESC

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('alvin kamara', 'RB', 13, 17.92, '8', '2023', 159458, 1, 2, 1, 6, 2, 1, 17.92, 13, 0, 13, 0),
 ('breece hall', 'RB', 17, 17.09, '5', '2023', 159458, 2, 3, 1, 6, 2, 1, 17.09, 30, 0, 17, 0),
 ('jahmyr gibbs', 'RB', 15, 16.14, '4', '2023', 159458, 3, 4, 1, 6, 2, 1, 0, 45, 1, 4, 11),
 ('rachaad white', 'RB', 17, 15.76, '7', '2023', 159458, 4, 5, 1, 6, 2, 1, 15.76, 62, 1, -11, 28),
 ('derrick henry', 'RB', 17, 14.51, '2', '2023', 159458, 5, 6, 1, 6, 2, 1, 0, 79, 1, -28, 45),
 ('khalil herbert', 'RB', 12, 9.38, '9', '2023', 159458, 6, 9, 1, 6, 2, 1, 0, 91, 1, -45, 57)]

the cell above is a preview for how we're going to aggregate team score by
total points and include injuries

we need to see
1. when the running game total for that position goes over the total for that position
1. when it does, we need to remove the games where that person was the starter
1. all of the remaining games are flexable games
1. then we do the same calculation over again using flex positions

In [276]:

# Number of items to choose from
n = 10

# Number of selections
r = 15

# Calculate the number of combinations with repetition
combinations_with_repetition = n ** r

print(f"The number of combinations is: {combinations_with_repetition:,}")
# This is the output ... 
# 1,000,000,000,000,000
# ... that's a quadrillion possibilities
# and we need a way to make that way way way way smaller

# Under the ppg framework, we limited it to only 4 players per round at a maximum because we selected the best player
ppg = 4**10
print(f"Using ppg {ppg:,}")
# at each position in that round
# but when we use total points, we can't really do that because we want to look at places where the PPG is not the same as total points scored
# we kind of need a way to create some combos, score them, eliminate some, and continue with some others
# for example, we can run through 3 rounds
# and remove anyone w/ two QBs or two TEs
# then proceed by adding other combos to those combos
# What chunk size shoudl we chose?
rd3 = 10**3
print(f"3 round chunks: {rd3:,}")
rd4 = 10**4
print(f"4 round chunks: {rd4:,}")
rd5 = 10**5
print(f"5 round chunks: {rd5:,}")

# It seems like we should do 5 round chunks
# Then we should disqualify some of the combos
# Where they have too many 

# Maybe we can slim this down by only looking at players that end up in the top 200? 
# That won't be all of the injured players though 

The number of combinations is: 1,000,000,000,000,000
Using ppg 1,048,576
3 round chunks: 1,000
4 round chunks: 10,000
5 round chunks: 100,000


In [279]:
%%sql
SELECT json('{"test": "value"}');


 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('{"test":"value"}',)]

In [307]:
%%sql
SELECT *
FROM image_ep_1_accuracy_graph_focus
WHERE 1=1
    AND pick_accuracy < 0
    AND (
        position='QB' AND draft_rank < 20
        OR position!='QB')
ORDER BY pick_accuracy ASC
LIMIT 20

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('2016', 'TE', 'david johnson', 2, 22.83, 1.06, 81, -21.77, 21.77, 1.06, 22.83, 'firebrick', 22.83, 'david johnson<br>TE2 to 81 (-21.77) ppg'),
 ('2016', 'RB', 'adrian peterson', 4, 19.94, 3.0, 95, -16.94, 16.94, 3.0, 19.94, 'firebrick', 19.94, 'adrian peterson<br>RB4 to 95 (-16.94) ppg'),
 ('2020', 'RB', 'saquon barkley', 2, 22.83, 7.7, 56, -15.13, 15.13, 7.7, 22.83, 'firebrick', 22.83, 'saquon barkley<br>RB2 to 56 (-15.13) ppg'),
 ('2016', 'RB', 'todd gurley', 1, 26.09, 12.39, 24, -13.7, 13.7, 12.39, 26.09, 'firebrick', 26.09, 'todd gurley<br>RB1 to 24 (-13.7) ppg'),
 ('2017', 'RB', 'david johnson', 1, 26.09, 13.0, 17, -13.09, 13.09, 13.0, 26.09, 'firebrick', 26.09, 'david johnson<br>RB1 to 17 (-13.09) ppg'),
 ('2016', 'RB', 'jamaal charles', 8, 17.52, 4.47, 78, -13.05, 13.05, 4.47, 17.52, 'firebrick', 17.52, 'jamaal charles<br>RB8 to 78 (-13.05) ppg'),
 ('2022', 'RB', 'jonathan taylor', 1, 26.09, 13.31, 18, -12.78, 12.78, 13.31, 26.09, 'firebrick', 26.09, 'jonathan taylor<br>RB1 to 18 (-12.78) ppg'),
 ('2022', 'QB', 'trey lance', 12, 18.31, 6.25, 56, -12.06, 12.06, 6.25, 18.31, 'firebrick', 18.31, 'trey lance<br>QB12 to 56 (-12.06) ppg'),
 ('2017', 'WR', 'allen robinson', 21, 14.25, 2.7, 136, -11.55, 11.55, 2.7, 14.25, 'firebrick', 14.25, 'allen robinson<br>WR21 to 136 (-11.55) ppg'),
 ('2020', 'WR', 'michael thomas', 1, 23.2, 11.99, 41, -11.21, 11.21, 11.99, 23.2, 'firebrick', 23.2, 'michael thomas<br>WR1 to 41 (-11.21) ppg'),
 ('2017', 'WR', 'willie snead', 35, 12.18, 1.38, 167, -10.8, 10.8, 1.38, 12.18, 'firebrick', 12.18, 'willie snead<br>WR35 to 167 (-10.8) ppg'),
 ('2019', 'QB', 'ben roethlisberger', 15, 17.03, 6.35, 48, -10.68, 10.68, 6.35, 17.03, 'firebrick', 17.03, 'ben roethlisberger<br>QB15 to 48 (-10.68) ppg'),
 ('2019', 'QB', 'cam newton', 10, 18.91, 8.35, 44, -10.56, 10.56, 8.35, 18.91, 'firebrick', 18.91, 'cam newton<br>QB10 to 44 (-10.56) ppg'),
 ('2017', 'RB', 'darren mcfadden', 39, 9.75, -0.2, 145, -9.95, 9.95, -0.2, 9.75, 'firebrick', 9.75, 'darren mcfadden<br>RB39 to 145 (-9.95) ppg'),
 ('2017', 'WR', 'terrelle pryor', 16, 15.23, 5.56, 96, -9.67, 9.67, 5.56, 15.23, 'firebrick', 15.23, 'terrelle pryor<br>WR16 to 96 (-9.67) ppg'),
 ('2023', 'RB', 'austin ekeler', 2, 22.83, 13.24, 21, -9.59, 9.59, 13.24, 22.83, 'firebrick', 22.83, 'austin ekeler<br>RB2 to 21 (-9.59) ppg'),
 ('2019', 'WR', 'donte moncrief', 50, 10.29, 0.73, 194, -9.56, 9.56, 0.73, 10.29, 'firebrick', 10.29, 'donte moncrief<br>WR50 to 194 (-9.56) ppg'),
 ('2021', 'RB', 'raheem mostert', 29, 11.5, 2.0, 117, -9.5, 9.5, 2.0, 11.5, 'firebrick', 11.5, 'raheem mostert<br>RB29 to 117 (-9.5) ppg'),
 ('2023', 'RB', 'dalvin cook', 26, 12.14, 2.68, 86, -9.46, 9.46, 2.68, 12.14, 'firebrick', 12.14, 'dalvin cook<br>RB26 to 86 (-9.46) ppg'),
 ('2017', 'RB', 'paul perkins', 33, 10.81, 1.96, 107, -8.85, 8.85, 1.96, 10.81, 'firebrick', 10.81, 'paul perkins<br>RB33 to 107 (-8.85) ppg')]

In [319]:
%%sql
SELECT
    c_year AS year,
    c_pos AS position,
    adp_pos_rank AS draft_rank,
    exp_fppg_ppr AS expected_ppg
FROM nfl_adp_hype
WHERE 1=1
    AND c_pos IN ("QB", "WR", "RB", "TE")
    AND fppg_ppr IS NOT NULL
    AND (
        (draft_rank IS NOT NULL AND expected_ppg IS NOT NULL)
        )
UNION ALL
SELECT
    'Avg.' AS year,
    fantasy_pos AS position,
    actual_pos_fin_fppg_ppr AS draft_rank,
    exp_fppg_ppr AS expected_ppg
FROM hist_adp_perf
LIMIT 20

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('2016', 'QB', 8, 19.28),
 ('2016', 'QB', 67, 0.82),
 ('2016', 'QB', 19, 16.06),
 ('2016', 'QB', 57, 2.95),
 ('2016', 'QB', 25, 14.54),
 ('2016', 'QB', 48, 7.22),
 ('2016', 'QB', 14, 17.22),
 ('2016', 'QB', 49, 6.69),
 ('2016', 'QB', 73, 0.15),
 ('2016', 'QB', 1, 25.09),
 ('2016', 'QB', 55, 3.77),
 ('2016', 'QB', 82, -0.25),
 ('2016', 'QB', 41, 9.97),
 ('2016', 'QB', 27, 14.17),
 ('2016', 'QB', 18, 16.31),
 ('2016', 'QB', 52, 4.97),
 ('2016', 'QB', 23, 15.02),
 ('2016', 'QB', 40, 10.2),
 ('2016', 'QB', 81, -0.2),
 ('2016', 'QB', 22, 15.23)]

In [320]:
%%sql
SELECT
    fantasy_pos AS position,
    count(1) 
FROM hist_adp_perf
GROUP BY 1
LIMIT 10

 * sqlite:////Users/beoconno/Documents/ff/ff_data/db_output/nfl.db
Done.


[('QB', 84), ('RB', 164), ('TE', 148), ('WR', 274)]